---
title: "Assessing the impact of sequencing characteristics on 16S rRNA marker-gene surveys beta diversity analysis."

output:
  # bookdown::pdf_document2:
  bookdown::word_document2:
    toc: FALSE
    # template: template.tex
    # keep_tex: TRUE
    # pandoc_args: [
    #   "--bibliography=bibliography.bib",
    #   "--biblatex"
    # ]
    # citation_package: biblatex
documentclass: "amsart"
bibliography: [bibliography.bib]
---

```{r include = FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE, echo = FALSE, fig.path = "img/", fig.width = 5)
source("src/plot_color_definitions.R")
```

<!-- 
BMC Microbiome Submission https://microbiomejournal.biomedcentral.com/submission-guidelines/preparing-your-manuscript/methodology 

Rmarkdown with BMC template https://stackoverflow.com/questions/44373540/no-author-details-in-pdf-from-r-markdown-latex-template 
-->

# Abstract
```{r introduction, child="docs/abstract.Rmd"}
``` 

# Introduction
```{r introduction, child="docs/introduction.Rmd"}
``` 


# Methods  
```{r methods, child="docs/methods.Rmd"}
``` 

# Results 
```{r experimentalDesign, echo = FALSE, fig.cap = "Stand-in figure."}
knitr::include_graphics("img/experimental_design.png")
```

The beta diversity assessment framework includes three components; (1) evaluating beta diversity between PCR replicates for sequencing runs with different error rates and library sizes, (2) difference in beta diversity between biological and technical replicates, and (3) ability to differentiate between PCR replicates from unmixed pre-exposure samples and post-exposure samples as well as the unmixed post-exposure samples.  
For our assessment we used a dataset consisting of two-sample titrations of DNA extracts from five vaccine trial participant stool samples collected before and after exposure to pathogenic _E. coli_ (Fig. \@ref(fig:experimentalDesign)). 

## Dataset Characteristics  

```{r seqPipeChar, child="reports/seq_and_pipe_characterization.Rmd"}
```

```{r sampleChar, child="reports/sample_characterization.Rmd", eval = FALSE}
## Excluding for now, not sure how it fits into the narrative
```

## PCR Repeatability
```{r techArt, child="reports/tech_artifacts.Rmd"}
```


## Signal to Noise

```{r compToExp, child="reports/signal_to_noise.Rmd"}
```


## Biological v. Technical Variation

```{r bioVtech, child="reports/biological_v_technical.Rmd"}
```


# Discussion
```{r discussion, child="docs/discussion.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
``` 

# References 
<!-- <div id="refs"></div> -->
