---
title: "Biological versus Technical Variability"
author: "Jackie Meisel"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2:
    toc: FALSE
---

```{r bioVtechSetup, warning=FALSE, message=FALSE, echo = FALSE}
library(ProjectTemplate)
load.project()
source("src/plot_color_definitions.R")
library(ggthemes)
library(ggpubr)
```



```{r bioVtechMunge, warning=FALSE, message=FALSE, echo = FALSE}
## Pre-processing data for figure(s)
bio_v_tech_dat <- biol_v_tech_variation[[3]] %>%
    mutate(f_type = if_else(metric %in% c("unifrac","wunifrac"),
                        "Phylogenetic","Taxonomic"),
            weight = if_else(metric %in% c("unifrac","jaccard"),
                        "Unweighted", "Weighted")) %>%
    mutate(pipe = case_when(pipe == "deblur" ~ "q_deblur",
                        pipe == "qiimeOpenRef" ~ "q_open",
                        pipe == "qiimeClosedRef" ~ "q_closed",
                        pipe == "qiimeDeNovo"~ "q_denovo",
                        TRUE ~ pipe),
           pipe = factor(pipe))


biol_v_tech_variation_summary_highlevel <- bio_v_tech_dat %>%
    group_by(variation, pipe, normalization, metric, normalization_type) %>%
    summarise(mean_value = mean(value),
              stdev = sd(value),
              N = n(),
              se = stdev/sqrt(N))

biol_v_tech_variation_summary_higherlevel <- biol_v_tech_variation_summary_highlevel[, 1:6] %>%
    spread(variation, mean_value) %>%
    mutate(mean_difference  = biological - technical) %>%
    mutate(f_type = if_else(metric %in% c("unifrac","wunifrac"),
                        "Phylogenetic","Taxonomic"),
            weight = if_else(metric %in% c("unifrac","jaccard"),
                        "Unweighted", "Weighted"))

```


```{r bioVtechPlot1, warning=FALSE, message=FALSE, echo = FALSE, fig.cap = "Biological vs. Technical Variation, y-axis is the differences between the mean biological (subject and titration level) and technical variation (sequencing lab and run) (pairwise distance between replicates.) Grey line indicates the mean differences for diversity metrics calculated using raw counts. Higher values indicate better differentiation between technical variability and true biological differences. Points indicate mean differences for diversity metrics calculated using normalized counts with color indicating normalization method.", fig.height=4, fig.width=6, eval = FALSE}

raw_bio_v_tech <- biol_v_tech_variation_summary_higherlevel %>%
    filter(normalization == "RAW") %>%
    ungroup()

biol_v_tech_variation_summary_higherlevel %>%
    filter(normalization != "RAW") %>%
    ggplot(aes(x = pipe)) +
        geom_blank() +
        geom_line(data = raw_bio_v_tech,
                  aes(x = as.numeric(pipe), y = mean_difference), color = "grey60") +
        geom_point(aes(x = pipe, y = mean_difference, fill = normalization),
                   shape = 21) +
        theme_bw() +
        facet_grid(f_type~weight, scales = "free_y") +
        theme(axis.text.x = element_text(angle = -45, hjust = 0),
              legend.position = "bottom") +
        labs(x = "Pipeline", y = "Biological - Technical", fill = "Normalization") +
    scale_fill_manual(values = norm_colors)

```


```{r bioVtechRaw, warning=FALSE, message=FALSE, echo = FALSE, fig.cap = ".Biological vs. Technical Variation,distribution in (A) weighted and (B) unweighted beta-diversity between between technical replicates and biological treatments (subject  and timepoint)."}

weighted <- bio_v_tech_dat %>%
    filter(weight == "Weighted", normalization == "RAW") %>%
    ggplot(aes(x = variation_label, y = value,
               color = variation, fill = variation)) +
    geom_tufteboxplot() +
    facet_grid(f_type~pipe, scales = "free_y") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -45, hjust = 0),
          legend.position = "bottom") +
    labs(x = "Variation Type", y = "Beta-Diversity",
         fill = "Variation Source", color = "Variation Source")


unweighted <- bio_v_tech_dat %>%
    filter(weight == "Unweighted", normalization == "RAW") %>%
    ggplot(aes(x = variation_label, y = value,
               color = variation, fill = variation)) +
    geom_tufteboxplot() +
    facet_grid(f_type~pipe, scales = "free_y") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -45, hjust = 0),
          legend.position = "bottom") +
    labs(x = "Variation Type", y = "Beta-Diversity",
         fill = "Variation Source", color = "Variation Source")

ggarrange(weighted + rremove("xlab") + rremove("x.text"),
          unweighted,
          ncol = 1, nrow = 2,
          labels = "AUTO", common.legend = TRUE,
          legend = "bottom", heights = c(0.35,0.65))
```

```{r bioVtechVarpart, warning=FALSE, message=FALSE, echo = FALSE, fig.cap = "Biological vs. Technical Variation, y-axis is the adjusted R^2 value, indicating proportion of variance explained by each biological (subject and titration) and technical (seq run) variable. Black bars represent values for unnormalized data.", fig.width = 6, fig.height = 8}
varpart_stats <- varpart_stats %>%
        mutate(pipe = case_when(pipe == "deblur" ~ "q_deblur",
                        pipe == "qiimeOpenRef" ~ "q_open",
                        pipe == "qiimeClosedRef" ~ "q_closed",
                        pipe == "qiimeDeNovo"~ "q_denovo",
                        TRUE ~ pipe),
           pipe = factor(pipe))

varpart_weighted <- varpart_stats %>%
    filter(effect == "conditional",
           metric %in% c("bray", "wunifrac"))

varpart_weighted_raw <- varpart_weighted %>% filter(normalization == "RAW")
varpart_weighted_norm <- varpart_weighted %>% filter(normalization != "RAW")

weighted <- ggplot(varpart_weighted_norm) +
    geom_bar(aes(x = feature, y = Adj.R.square, fill = normalization),
             stat = "identity", position = "dodge", width = 0.5) +
    geom_point(data = varpart_weighted_raw,
               aes(x = feature, y = Adj.R.square),
               shape = "-", size = 8) +
    facet_grid(metric~pipe, scales = "free_x", space = "free") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -45, hjust = 0)) +
    xlab("") + ylab("Adjusted R^2") +
    theme(legend.position = "bottom") +
    scale_fill_manual(values = norm_colors)

varpart_unweighted <- varpart_stats %>%
    filter(effect == "conditional",
           metric %in% c("jaccard", "unifrac"))

varpart_unweighted_raw <- varpart_unweighted %>% filter(normalization == "RAW")
varpart_unweighted_norm <- varpart_unweighted %>% filter(normalization != "RAW")

unweighted <- ggplot(varpart_unweighted_norm) +
    geom_bar(aes(x = feature, y = Adj.R.square, fill = normalization),
             stat = "identity", position = "dodge", width = 0.5) +
    geom_point(data = varpart_unweighted_raw,
               aes(x = feature, y = Adj.R.square),
               shape = "-", size = 8) +
    facet_grid(metric~pipe, scales = "free_x", space = "free") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -45, hjust = 0)) +
    xlab("") + ylab("Adjusted R^2") +
    theme(legend.position = "bottom") +
    scale_fill_manual(values = norm_colors)

ggarrange(weighted, unweighted,
          ncol = 1, nrow = 2,
          labels = "AUTO",common.legend = TRUE,
          legend = "bottom")
```


Finally, we characterized how different pipelines and normalization methods capture diversity differences between biological factors and technical replicates.
As expected, the mean diversity observed between biological factors was greater than between technical replicates (Fig. \@ref(fig:bioVtechRaw)).
The magnitude of this difference, however, was greater for weighted than unweighted beta-diversity metrics and varied by pipeline.
Greater differences were observed with the DADA2, Mothur, and Deblur pipelines, compared to the QIIME clustering approaches.


Variation partitioning was used to identify the amount of variation attributable to subject, titration factor (unmixed pre-exposure and unmixed post-exposure), and sequencing run.
Across all pipelines and diversity metrics, the greatest amount of variation is often explained by subject, followed by titration factor (Fig. \@ref(fig:bioVtechVarpart)).
This is consistent with our observation of greater biological than technical variability. Sequencing run accounts for a greater proportion of the explained variance in the unnormalized runs, highlighting the overall importance of normalizing our datasets.


Effective normalization methods decrease technical noise in the data without decreasing biological signal.
For both weighted (Fig. \@ref(fig:bioVtechVarpart)A) and unweighted (Fig. \@ref(fig:bioVtechVarpart)B) metrics, rarefaction normalization methods generally show increased proportion of variation explained by biological factors and decreased proportion of variation explained by technical artifacts.
Numeric normalization methods were not as effective, especially for the QIIME pipelines. RLE and TMM normalization consistently increased technical variability and often decreased biological variability (Fig. \@ref(fig:bioVtechVarpart)A).
