---
title: "Biological versus Technical Variability"
author: "Jackie Meisel"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2: 
    toc: FALSE
---

<!--
Biological versus technical variation results section, comparison of pairwise distances between biological and technical replicates and partitioning total variation into biological and technical components. 
-->

```{r bioVtechSetup, warning=FALSE, message=FALSE, echo = FALSE}
library(ProjectTemplate)
load.project()
## load required libraries
```

```{r bioVtechMunge, warning=FALSE, message=FALSE, echo = FALSE}
## Pre-processing data for figure(s)
biol_v_tech_variation_summary_highlevel <- biol_v_tech_variation[[3]] %>% 
    group_by(variation, pipe, normalization, metric, normalization_type) %>%
    summarise(mean_value = mean(value), 
              stdev = sd(value), 
              N = n(), 
              se = stdev/sqrt(N))

biol_v_tech_variation_summary_higherlevel <- biol_v_tech_variation_summary_highlevel[, 1:6] %>%
    spread(variation, mean_value) %>%
    mutate(mean_difference  = biological - technical) %>% 
    mutate(f_type = if_else(metric %in% c("unifrac","wunifrac"), 
                        "Phylogenetic","Taxonomic"),
            weight = if_else(metric %in% c("unifrac","jaccard"),
                        "Unweighted", "Weighted"))
```

```{r bioVtechPlot1, warning=FALSE, message=FALSE, echo = FALSE, fig.cap = "Biological vs. Technical Variation, y-axis is the differences between the mean biological and technical variation (pairwise distance between replicates.) Grey line indicates the mean differences for diversity metrics calculated using raw counts. Higher values indicate better differentiation between technical variability and true biological differences. Points indicate mean differences for diversity metrics calculated using normalized counts with color indicating normalization method."}
raw_bio_v_tech <- biol_v_tech_variation_summary_higherlevel %>% filter(normalization == "RAW") %>% 
    ungroup() %>% 
    mutate(pipe = factor(pipe))

biol_v_tech_variation_summary_higherlevel %>% 
    filter(normalization != "RAW") %>% 
    ggplot(aes(x = pipe)) + 
        geom_blank() + 
        geom_line(data = raw_bio_v_tech, 
                  aes(x = as.numeric(pipe), y = mean_difference), color = "grey60") + 
        geom_point(aes(x = pipe, y = mean_difference, fill = normalization),
                   shape = 21) + 
        theme_bw() + 
        facet_grid(f_type~weight, scales = "free_y") + 
        theme(axis.text.x = element_text(angle = -45, hjust = 0),
              legend.position = "bottom") + 
        labs(x = "Pipeline", y = "Biological - Technical", fill = "Normalization")
```

```{r bioVtechPlot2, warning=FALSE, message=FALSE, echo = FALSE, fig.cap = "Biological vs. Technical Variation, y-axis is the adjusted R^2 value, indicating proportion of variance explained by each biological (subject and titration) and technical (seq run) variable"}
varpart_trim <- varpart_stats %>% filter(effect == "conditional", 
                                         normalization %in% c("RAW","rare5000", "rareq15","RLE","TMM"), 
                                         metric %in% c("bray", "wunifrac")) %>% 
    mutate(normalization = factor(normalization, 
                                  levels = c("RAW", "rare5000", "rareq15", "RLE", "TMM"), ordered = T))

varpart_trim %>% filter(normalization != "RAW") %>% 
ggplot() +
    geom_bar(aes(x = feature, y = Adj.R.square, fill = normalization),
             stat = "identity", position = "dodge", width = 0.5) + 
    geom_point(data = varpart_trim %>% filter(normalization == "RAW"),
               aes(x = feature, y = Adj.R.square)) + 
    theme_bw() +
    theme(axis.text.x = element_text(angle = -45, hjust = 0)) +
    xlab("") + ylab("Adjusted R^2") + 
    facet_grid(metric~pipe, scales = "free_x", space = "free") +
    theme(legend.position = "bottom")
``` 
- Effective normalization methods decrease the technical variability in the data without decreasing the biological variability. 
- RLE and TMM consistently increase the technical variability and often decrease the amount of variability in the data attributed to biological factors.  

__Key Points__  

Beta-diversity distances between biological and technical replicates varies by pipeline and beta-diversity metric (Fig. \@ref(fig:bioVtechPlot1)).  

* Generally, greater differences between biological and technical replicates were observed using taxonomic metrics, rather than pylogenetic metrics (with the exception of mothur for weighted phylogenetic metrics).

* For weighted metrics, RLE and TMM decreased the difference relative to raw counts indicating that for beta diversity analysis these normalization methods reduce the power to distingush true biological differences from technical variability or noise.  

*In general, rarified count data had higher mean difference between biological and technical variation (excluding when rarefying to 15th quantile for qiime de novo, closed , and open-reference, which is potentially due to sample loss). 

Generally, biological features (subject and titration factor) explain most of the observed variance (Fig. \@ref(fig:bioVtechPlot1)).  

* Un-normalized (raw) datasets often showed greater influence of technical factors (seq run)

* Non-rarefaction normalization metrics (esp. RLE and TMM) showed greater influence of technical factors (seq run)-- especially for QIIME pipelines

__Very very very rough draft of text__  

Beta-diversity distances between biological and technical replicates varies by pipeline and beta-diversity metric

We next looked at how different pipelines and normalization methods captured diversity differences between our biological and technical replicates. Overall, as expected, the mean diversity observed between biological replicates was greater than that between technical replicates. Generally, taxonomic diversity metrics were better at capturing the biological signal than pylogenetic metrics. One exception was that the weighted unifrac metric mirrored the trends observed with taxonomic metrics when using the mothur pipeline. 

For unweighted metrics, normalization by rarefaction produced consistent differences in variation accross most subsampling depths. The three QIIME pipelines, however, identified smaller differences between biological and technical replicates when subsampled to the 15th quantile. This inconsistency is most likely due to greater sample loss in these pipelines at this subsampling level [?? is this true]. Amongst the non-rarefaction normalization methods, RLE and TMM most noticeably reduced the power to distingush biological signal from technical variability.

We also used variation partitioning to determine the amount of variation attributable to subject, titration factor (unmixed pre-exposure and unmixed post-exposure), and sequencing run. Accross all pipelines and diversity metrics, the greatest amount of variation is often explained by subject, followed by titration factor. In our unnormalized pipelines, sequencing run accounts for a greater proprotion of the explained variance, highlighting the overall importance of normalizing our datasets. Rarefaction normalization methods generally show increased amounts of variation explained by biological factors rather than technical artifacts. The non-rarefaction normalization methods do not reduce the impacts of technical arifacts as effectively, especially for the QIIME pipelines. Again, the TMM and RLE methods were consistently less effective than other normalization methods.
