---
title: "Sequence Data and Pipeline Characterization"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2: 
    toc: FALSE
---

```{r seqPipeCharSetup, warning=FALSE, message=FALSE, echo = FALSE}
library(ProjectTemplate)
load.project()
library(ggpubr)
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```


## Dataset Characterization - Figure 1
* Figure 1A: Sequence quality - highlighting sequencing run differences   

```{r phix_error_fig, fig.cap = "PhiX error rate relative to 16S rRNA amplicon base position for initial and the four sequencing runs."}
ggplot(phix_error_df) +
    geom_point(aes(x = base_position, y = errorrate, color = ds), alpha = 0.01) +
    geom_smooth(aes(x = base_position, y = errorrate, color = ds, linetype = read)) +
    theme_bw() +
    labs(x = "Amplicon Base Position", y = "PhiX Error Rate", color = "Dataset")
```

* Supplemental - DADA2 seq quality plot for same PCR replicate across runs 

* Figure 1B: Reads per sample - relevant to normalization 

- Good separation between sample and no template control library size for JHU but not NIST samples.  
- Total abundance is lower for samples compared to no template controls for most sequencing runs and samples. Though a few no template controls have values within the sample range.  (see pipe characterization total abundance and pass rate plots). 

```{r fig_count, fig.cap="Distribution in the number of reads per barcoded sample (Library Size) by sequncing laboratory and sequencing run. Negative control library size is indicated by hollow points. Negative controls were not included in the boxplots."}
r1_seq_char_df <- seq_char_df %>%
    filter(read_dir == "R1") %>%
    left_join(mgtstMetadata) %>%
    mutate(seq_run = as.character(seq_run))
r1_ntc <- r1_seq_char_df %>% filter(biosample_id == "NTC") 

r1_seq_char_df %>% filter(biosample_id != "NTC") %>%
    ggplot() + 
        geom_boxplot(aes(x = seq_lab, y = read, color = seq_run)) + 
        geom_point(data = r1_ntc, aes(x = seq_lab, y = read, color = seq_run), 
                   shape = 21,
                   position = position_jitterdodge(jitter.width = 0.25)) + 
        scale_y_log10() + 
        theme_bw() + labs(x = "Sequencing Lab", y = "Library Size", color = "Sequencing Run") +
        theme(legend.position = "bottom")
```






## Pipeline characterization   

- Pipelines varied in terms of the number of features, total abundance, and proportion of sequences passing quality control.  

```{r pipeCharTbl, echo = FALSE, message = FALSE, warning = FALSE}
pipe_char_summary <- pipe_char_df %>%
    unnest() %>% 
    left_join(mgtstMetadata) %>% 
    filter(biosample_id != "NTC") %>% 
    group_by(pipe, n_taxa, n_samples, sparsity) %>% 
    summarise(total_abu_med = round(median(total_abu),0),
              total_abu_max = round(max(total_abu),0),
              total_abu_min = round(min(total_abu),0), 
              pass_rate_med = round(median(pass_rate),2),
              pass_rate_max = round(max(pass_rate),2),
              pass_rate_min = round(min(pass_rate),2)) %>% 
    mutate(total_abu_summary = paste0(total_abu_med, " (", total_abu_max, "-", total_abu_min,")")) %>% 
    mutate(pass_rate_summary = paste0(pass_rate_med, " (", pass_rate_max, "-", pass_rate_min,")")) %>% 
    mutate(sparsity = round(sparsity, 3)) %>% 
    select(-total_abu_med, -total_abu_max, -total_abu_min,
           -pass_rate_med, -pass_rate_max, -pass_rate_min)

pipe_char_summary %>% 
    select(pipe, n_taxa, n_samples, sparsity, total_abu_summary, pass_rate_summary) %>% 
      dplyr::rename(Pipelines = pipe, Features = n_taxa, Samples = n_samples, 
                    Sparsity = sparsity, 
                    `Total Abundance` = total_abu_summary, `Pass Rate` = pass_rate_summary) %>% 
      knitr::kable(caption = "Summary statistics for the different bioinformatic pipeliens. Four pipelines, de novo, open reference, closed reference, and deblur (sequence inference), used the sample sequence pre-processing methods. DADA2 is a denoising sequence inference pipeline and mothur is a de-novo clustering pipeline. No template controls were excluded from summary statistics. Sparsity is the proportion of 0's in the count table. Features is the total number of OTUs (QIIME and mothur) or SVs (DADA2) in the count. Sample coverage is the median and range (minimum - maximum) per sample total feature abundance. Filter rate is the proportion of reads that were removed while processing the sequencing data for each bioinformatic pipeline.", booktabs = TRUE)
```

## Figure 2 
```{r}
ps_list <- list.files("data/phyloseq_objects", full.names = TRUE) %>% 
    set_names(basename(.)) %>% 
    map(readRDS)
```

* Species abundance curves - relate to ecology literature, the species abundance curves are fundamentally different between pipelines, indicating that the features generated using represent different biological units and therefore should be interpreted as such (Only unmixed samples same PCR replicate across sequencing runs).  

```{r}
get_rad_df <- function(ps, sample_name) {
    x_abu <- prune_samples(sample_name, ps) %>%
        {prune_taxa(taxa_sums(.) > 0, .)} %>%
        otu_table() %>%
        as.vector() %>%
        sort(decreasing = TRUE)
    
    data_frame(rank = 1:length(x_abu), abund = x_abu)
}

rad_df <- ps_list %>% map_df(get_rad_df, "nist_run1_1-A1", .id = "pipe")

ggplot(rad_df) + geom_point(aes(x = rank, y = abund)) + 
    scale_y_log10() + 
    theme_bw() + 
    labs(x = "Feature Rank", y = "Abundance") + 
    facet_wrap(~pipe, scales = "free_x")
```

* Rarefaction curves (Only unmixed samples same PCR replicate across sequencing runs)
```{r}
ps_iNEXT <- function(ps){
    sample_set <- c("nist_run1_2-A10","nist_run2_2-A10",
                    "jhu_run1_2-A10","jhu_run2_2-A10")
    # sample_set <- c("nist_run1_2-A10","nist_run2_2-A10")
    count_tbl <- prune_samples(sample_set, ps)  %>%
        {prune_taxa(taxa_sums(.) > 0, .)} %>%
        otu_table() 
    
    if (!taxa_are_rows(ps)) {
        count_tbl <- t(count_tbl)
    } 

    count_df <- as.data.frame(count_tbl)
    
    iNEXT::iNEXT(count_df)
}

ps_list %>% map(ps_iNEXT) %>% walk(~{plot(iNEXT::ggiNEXT(.) + theme_bw())})
```


* Prevalence (all samples)
```{r}
## Code modified from Callahan F1000
get_prevdf <- function(ps){
    prevdf <- apply(X = otu_table(ps), 
                MARGIN = ifelse(taxa_are_rows(ps), yes = 1, no = 2),
                FUN = function(x){sum(x > 0)})

    # Add taxonomy and total read counts to this data.frame
    data.frame(Prevalence = prevdf, 
               TotalAbundance = taxa_sums(ps), 
               tax_table(ps)) %>% 
        mutate(PrevFrac = Prevalence / nsamples(ps))
}


pipe_prev_df <- ps_list %>% map_df(get_prevdf, .id = "pipe")

ggplot(pipe_prev_df, aes(TotalAbundance, PrevFrac)) +
    geom_point(size = 2, alpha = 0.7) +
    scale_x_log10() +
    labs(x = "Total Abundance", y = "Prevalence [Frac. Samples]") +
    facet_wrap( ~ pipe) + theme_bw()
```





## Supplemental Sequencing Data Characterization  
```{r}
fqs <- list.files("data/example_seq_data", full.names = TRUE)
dada2::plotQualityProfile(fqs)
```


## No Template Control Investigation 
```{r}
pipe_char_meta_df <- pipe_char_df %>%
    unnest() %>% 
    left_join(mgtstMetadata) %>%
    mutate(seq_run = as.character(seq_run)) 

pipe_char_ntc <- pipe_char_meta_df %>% filter(biosample_id == "NTC") 

pipe_char_meta_df %>% filter(biosample_id != "NTC") %>%
    ggplot() + 
        geom_boxplot(aes(x = seq_lab, y = total_abu + 1, color = seq_run)) + 
        geom_point(data = pipe_char_ntc, aes(x = seq_lab, y = total_abu + 1, color = seq_run), 
                   shape = 21,
                   position = position_jitterdodge(jitter.width = 0.25)) + 
        scale_y_log10() + 
        theme_bw() + labs(x = "Sequencing Lab", y = "Total Abundance", color = "Sequencing Run") +
        theme(legend.position = "bottom") + facet_wrap(~pipe)
```

- Pass filter rate is lower for samples compared to no template controls excluding outliers. 
```{r}
pipe_char_meta_df %>% 
    mutate(biosample_id = if_else(biosample_id == "NTC", "NTC","Sample")) %>% 
    ggplot() + 
    geom_boxplot(aes(x = seq_run,  y = pass_rate, color = biosample_id)) + 
    facet_wrap(pipe~seq_lab, scales = "free_y") + 
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 90)) +
    labs(x = "Sequencing Run", y = "Pass Quality Control", color = "Sample Type")
```

No template control with high pass rate - all the same sample. 
```{r}
pipe_char_meta_df %>% filter(biosample_id == "NTC", pass_rate > 0.2)
```

Need to figure out better way to look at samples with low pass rate.
```{r}
pipe_char_meta_df %>% filter(biosample_id != "NTC") %>% 
    group_by(pipe, seq_lab, seq_run) %>% 
    top_n(n = 1, wt = -pass_rate)
```