---
title: "Sequence Data and Pipeline Characterization"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2:
    toc: FALSE
---

```{r seqPipeCharSetup, warning=FALSE, message=FALSE, echo = FALSE}
library(ProjectTemplate)
load.project()
source("src/plot_color_definitions.R")
ps_list <- list.files("data/phyloseq_objects", full.names = TRUE) %>%
    set_names(str_remove(basename(.), "_ps.rds")) %>%
    map(readRDS)

library(ggpubr)
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

```{r seqCharError}
error_plot <- ggplot(phix_error_df) +
    geom_point(aes(x = base_position, y = errorrate, color = ds), alpha = 0.01) +
    geom_smooth(aes(x = base_position, y = errorrate, color = ds, linetype = read)) +
    theme_bw() +
    labs(x = "Amplicon Base Position", y = "PhiX Error Rate", color = "Dataset") +
    scale_color_manual(values = run_colors) +
    theme(legend.direction = "horizontal",
          # legend.position = "bottom", legend.box = "vertical")
          legend.position = "bottom")
```

```{r seqCharCount}
r1_seq_char_df <- seq_char_df %>%
    filter(read_dir == "R1") %>%
    left_join(mgtstMetadata) %>%
    mutate(seq_run = as.character(seq_run))
r1_ntc <- r1_seq_char_df %>% filter(biosample_id == "NTC")

n_read_plot <- r1_seq_char_df %>% filter(biosample_id != "NTC") %>%
    ggplot() +
        geom_boxplot(aes(x = seq_lab, y = read, color = seq_run)) +
        geom_point(data = r1_ntc, aes(x = seq_lab, y = read, color = seq_run),
                   shape = 21,
                   position = position_jitterdodge(jitter.width = 0.25)) +
        scale_y_log10() +
        theme_bw() + labs(x = "Sequencing Lab",
                          y = "Library Size",
                          color = "Sequencing Run") +
        theme(legend.position = "bottom")# +
        # scale_color_manual(values = run_colors)
```


```{r covFit}
## TODO move to munge
qual_df <- seq_qual_comparisons %>%
    mutate(read_dir = paste0(read_dir, "_qual")) %>%
    dplyr::select(-sd_quality, -cov_quality) %>%
    spread(read_dir, mean_quality) %>%
    separate(replicate, c("biosample_id","t_fctr","seq_run_id"),
             remove = FALSE,extra = "merge")

tech_art_df <- seq_char_comparisons %>%
    dplyr::select(replicate,  pipe, normalization, metric, cov_num_reads, cov_total_abu) %>%
    left_join(qual_df)
fit2 <- lme4::lmer(cov_total_abu ~ seq_run_id + (1 | pipe / biosample_id / t_fctr),
                   data = tech_art_df)

confint_dat <- confint(fit2)
mod_betas <- fit2@beta
cov_est <- data_frame(run_id = c("jhu_run1","jhu_run2","nist_run1","nist_run2"),
                      est = c(mod_betas[1],
                              mod_betas[1] + mod_betas[2],
                              mod_betas[1] + mod_betas[3],
                              mod_betas[1] + mod_betas[4]))

fit_est_df <- as.data.frame(confint_dat) %>%
    rownames_to_column(var = "run_id") %>%
    filter(!grepl("sig", run_id)) %>%
    gather("conf_int","est", -run_id) %>%
    spread(run_id, est) %>%
    dplyr::rename(jhu_run1 = `(Intercept)`) %>%
    gather("run_id","est", -conf_int, -jhu_run1) %>%
    mutate(est = jhu_run1 + est) %>%
    mutate(run_id = str_replace(run_id, "seq_run_id","")) %>%
    spread(run_id, est) %>%
    gather("run_id","est", -conf_int) %>%
    mutate(conf_int = if_else(conf_int == "2.5 %", "lci","uci")) %>%
    spread(conf_int, est) %>%
    left_join(cov_est) %>%
    mutate(seq_run_id = str_replace(run_id, "_run",""))

cov_plot <- ggplot(fit_est_df) +
    geom_pointrange(aes(x = seq_run_id, y = est, ymin = lci,
                        ymax = uci, color = seq_run_id)) +
    labs(x = "Sequencing Run", y = "Total Abundance COV") +
    theme_bw() +
    scale_color_manual(values = run_colors) +
    theme(legend.position = "none")
```

```{r qualFit}
seq_qual_dat <- seq_char_df %>%
    left_join(mgtstMetadata) %>%
    filter(biosample_id != "NTC") %>%
    mutate(seq_run_id = paste0(seq_lab, seq_run))

## See 2018-02-15 for linear model results - all statistically significantly different

seq_qual_plot <- seq_qual_dat %>%
    ggplot() +
    geom_boxplot(aes(x = seq_run_id, y = quality, color = read_dir))  +
    theme_bw() +
    labs(x = "Sequencing Run", y = "Mode Read Base Quality Score", color = "Read Direction") +
    theme(legend.position = "right") +
    scale_color_manual(values = c("black", "darkgray"))
```


```{r seqCharPlot, fig.cap = "Sequencing quality and sample total abundance variation for the four sequencing runs used in this study. The same set of 192 PCRs were sequenced in all four runs. Independent sequencing libraries were generated at the two sequencing laboratories (JHU and NIST). (A) PhiX error rate relative to 16S rRNA amplicon base position for the four sequencing runs. (B) Distribution of mode read quality score by sequencing run. (C) Sequencing run total abundance coefficient of variation estimate and 95\\% confidence interval calculated using a mixed effects linear model.", fig.height=4, fig.width=10}
ggarrange(error_plot,
          ggarrange(seq_qual_plot, cov_plot, ncol = 2, labels = c("B", "C")), 
          nrow = 2, labels = "A") 

# seqCharPlot <- ggarrange(cov_plot, error_plot, seq_qual_plot, labels = "AUTO", ncol = 3)
# seqCharPlot
```

```{r seqCharPre, eval = FALSE, echo = FALSE}
ggarrange(seq_qual_plot, cov_plot)

tmp_img <- tempfile(fileext = ".png") 
ggsave(filename = tmp_img, device = "png",width = 8, height = 4)
img_filename <- "seq_char_plot.png"
## Add if statement to upload image for first time then update if already present
googledrive::drive_upload(media = tmp_img, type = "png",
             path =  "defense_presentation",
             name = img_filename)
# googledrive::drive_update(paste0("defense_presentation/", img_filename), tmp_img)
```


```{r pipeCharTbl, echo = FALSE, message = FALSE, warning = FALSE}
pipe_char_summary <- pipe_char_df %>%
    unnest() %>%
    mutate(pipe = if_else(pipe == "deblur", "q_deblur",
                          if_else(pipe == "qiimeOpenRef", "q_open",
                                  if_else(pipe == "qiimeClosedRef", "q_closed",
                                          if_else(pipe == "qiimeDeNovo", "q_denovo",
                                          pipe)))),
           pipe = factor(pipe)) %>%
    left_join(mgtstMetadata) %>%
    filter(biosample_id != "NTC") %>%
    group_by(pipe, n_taxa, n_samples, num_singletons, sparsity) %>%
    summarise(total_abu_med = round(median(total_abu),0),
              total_abu_max = round(max(total_abu),0),
              total_abu_min = round(min(total_abu),0),
              pass_rate_med = round(median(pass_rate),2),
              pass_rate_max = round(max(pass_rate),2),
              pass_rate_min = round(min(pass_rate),2)) %>%
    mutate(total_abu_summary = paste0(total_abu_med, " (",
                                      total_abu_max, "-", total_abu_min,")")) %>%
    mutate(pass_rate_summary = paste0(pass_rate_med, " (",
                                      pass_rate_max, "-", pass_rate_min,")")) %>%
    mutate(sparsity = round(sparsity, 3)) %>%
    dplyr::select(-total_abu_med, -total_abu_max, -total_abu_min,
           -pass_rate_med, -pass_rate_max, -pass_rate_min)

pipe_char_summary %>%
    dplyr::select(pipe, n_taxa, num_singletons, n_samples,  sparsity,
           total_abu_summary, pass_rate_summary) %>%
      dplyr::rename(Pipelines = pipe, Features = n_taxa,
                    Singletons = num_singletons,
                    Samples = n_samples,
                    Sparsity = sparsity,
                    `Total Abundance` = total_abu_summary,
                    `Pass Rate` = pass_rate_summary) %>%
      knitr::kable(caption = "Summary statistics for the different bioinformatic pipelines. No template controls were excluded from summary statistic calculations. Sparsity is defined as the proportion of 0's in the count table. Features is the total number of OTUs (QIIME and mothur) or SVs (DADA2), rows in the count table. Singletons is the total number of features only observed once in a single sample. Total Abundance is the median and range (minimum-maximum) per sample total feature abundance. Pass Rate is the median and range for the proportion of reads not removed while processing a sample's sequence data through a bioinformatic pipeline.", booktabs = TRUE)
```


```{r rareCurve}

## Moved code to munge 04.1_make_rare_df.R
# ps_iNEXT <- function(ps){
#     # sample_set <- c("nist_run1_2-A10","nist_run2_2-A10",
#     #                 "jhu_run1_2-A10","jhu_run2_2-A10")
#     seq_lab <- sample_data(ps)$seq_lab 
#     seq_run <- sample_data(ps)$seq_run
#     sample_data(ps)$seq_lab_run <- paste0(seq_lab, seq_run)
#     ps <- merge_samples(ps, group = "seq_lab_run")
#     
#     
#     
#     count_tbl <- ps %>% 
#         # prune_samples(sample_set, ps)  %>%
#         {prune_taxa(taxa_sums(.) > 0, .)} %>%
#         otu_table()
# 
#     if (!taxa_are_rows(ps)) {
#         count_tbl <- t(count_tbl)
#     }
# 
#     count_df <- as.data.frame(count_tbl)
# 
#     iNEXT::iNEXT(count_df)
# }
# 
# 
# rare_plot_df <- tibble(ps_obj = ps_list) %>%
#     add_column(pipe = names(ps_list)) %>%
#     mutate(pipe = case_when(pipe == "deblur" ~ "q_deblur",
#                         pipe == "qiimeOpenRef" ~ "q_open",
#                         pipe == "qiimeClosedRef" ~ "q_closed",
#                         pipe == "qiimeDeNovo"~ "q_denovo",
#                         TRUE ~ pipe),
#            pipe = factor(pipe)) %>%
#     mutate(inext_dat = map(ps_obj, ps_iNEXT),
#            rare_plot = map2(inext_dat, pipe, ~{iNEXT::ggiNEXT(.x) +
#                    labs(x = "Sampling Depth", y = "Feature Diversity") +
#                    theme_bw() +
#                    ggtitle(.y) +
#                    scale_shape_manual(values = run_shapes) +
#                    scale_color_manual(values=run_colors2)})) %>%
#     dplyr::select(-ps_obj, -inext_dat)

```

```{r rarePlot, fig.cap = "Rarefaction curves for an example sample across pipelines (A-F) and sequencing runs (line color). Rarefaction curves indicate how thoroughly a population has been sampled. Curves show the relationship between the number of unique features (y-axis) and sampling depth. Curves reaching an asymptote indicate the population has been completely sampled. Solid lines represent interpolated values obtained by randomly subsampling the observed abundance data. Dashed lines indicate extrapolated values predicted based on the observed count data and interpolated values.", fig.height=6, fig.width=8}
ggarrange(plotlist = rare_plot_df$rare_plot,
          labels = "AUTO", legend = "right",
          common.legend = TRUE, nrow = 3, ncol = 2)
```


The four replicate sequencing runs were of variable sequence quality and depth (Fig. \@ref(fig:seqCharPlot)).
Greater variability in sample feature total abundance was observed on the first run at each laboratory (Fig. \@ref(fig:seqCharPlot)A).
Sequencing error rates and base quality scores also varied by sequencing run.
JHU1 had higher PhiX error rates compared to all other runs, especially for the reverse reads (Fig. \@ref(fig:seqCharPlot)B).
Read base quality was lower for the reverse read than the forward reads for all four sequencing runs (Fig. \@ref(fig:seqCharPlot)C).
Sequence data from the two NIST runs had higher quality scores than the data from JHU runs, except for JHU2 forward reads (Fig. \@ref(fig:seqCharPlot)C).


Overall, sequences from JHU1 had lower read quality and higher variability in total sample abundance.
Sequences from NIST1 were of higher quality but also exhibited greater variability in total sample abundance.
Thus, by comparing the JHU1 results to the higher quality, less variable NIST2 and JHU2 runs, we can evaluate how well the bioinformatic pipelines handle low quality reads.
Similarly, we can use data from the NIST1 to determine how well normalization methods can account for differences in total abundance between samples.


Samples from the different sequencing runs were processed using six different bioinformatic pipelines.
Four of the pipelines, including the QIIME _de novo_, QIIME closed-reference, QIIME open-reference, Mothur _de novo_, utilize OTU clustering methods, while the remaining two, QIIME Deblur and DADA2, use sequence inference approaches.
Aside from the four QIIME pipelines each pipeline employs its own pre-processing, feature inference, and quality filtering methods.
The four QIIME pipelines used the same pre-processing methods.
As a result, the features and count tables generated by the pipelines exhibit different characteristics in terms of the number of features, total abundance, number of singletons, the proportion of sequences passing quality control (Table \@ref(tab:pipeCharTbl)).


<!-- Feature total abundance is the sum of all cells in the count table for a sample.
Differences in feature total abundance are due to differences in the number of reads generated per sample and proportion of sequences that pass the bioinformatic pipeline quality filter. -->


We generated rarefaction curves to assess feature diversity at multiple sampling depths for across the four sequencing runs (Fig. \@ref(fig:rarePlot)).
Sequence inference methods (DADA2 and Deblur) had lower overall feature diversity estimates and their rarefaction curves reached an asymptote around the same level (Fig. \@ref(fig:rarePlot)A & B), suggesting that sampling depth was sufficient to capture community diversity.
The JHU1 rarefaction curves at the origin for the QIIME pipelines was due to limited number of features, none for Deblur, were produced by the pipelines.
DADA2 asymptotes, however, were inconsistent across sequencing runs, indicating artificial plateaus for the lower throughput and lower quality runs (Fig. \@ref(fig:rarePlot)A).
Rarefaction curves for _de novo_, open-reference, and closed-reference methods did not reach an asymptote (Fig. \@ref(fig:rarePlot)).
The QIIME _de novo_ pipeline had the greatest slope, suggesting the highest rate of artifacts (Fig. \@ref(fig:rarePlot)E).
This is most likely due to the fact that the QIIME _de novo_ pipeline does not filter out singletons (Table \@ref(tab:pipeCharTbl)).
Furthermore, the Mothur rarefaction curves were consistent across sequencing runs, but the QIIME clustering pipelines rarefaction curves were influenced by both sequence quality and library size (Fig. \@ref(fig:rarePlot)D-F).


<!-- Rarefaction curves are used in ecology to determine how well a community has been sampled [@Gotelli2001;@Chao2014].
Measurement methods prone to errors, such as marker-gene sequencing, will never reach the asympotote if errors are not appropriately accounted for in sample processing [@Chiu2016]. -->

```{r rarePresFig, eval = FALSE, echo = FALSE}
## Plots for defense 
rare_plot_df <- tibble(ps_obj = ps_list) %>%
    add_column(pipe = names(ps_list)) %>%
    filter(pipe %in% c("dada", "mothur", "qiimeOpenRef")) %>% 
    mutate(pipe = case_when(pipe == "dada" ~ "DADA2",
                        pipe == "qiimeOpenRef" ~ "QIIME",
                        pipe == "mothur" ~ "Mothur"),
           pipe = factor(pipe)) %>%
    mutate(inext_dat = map(ps_obj, ps_iNEXT),
           rare_plot = map2(inext_dat, pipe, ~{iNEXT::ggiNEXT(.x) +
                   labs(x = "Sampling Depth", y = "Feature Diversity") +
                   theme_bw() +
                   ggtitle(.y) +
                   scale_shape_manual(values = run_shapes) +
                   scale_color_manual(values=run_colors2)})) %>%
    dplyr::select(-ps_obj, -inext_dat)

ggarrange(plotlist = rare_plot_df$rare_plot, legend = "bottom",
          common.legend = TRUE, nrow = 1, ncol = 3)

tmp_img <- tempfile(fileext = ".png") 
ggsave(filename = tmp_img, device = "png",width = 10, height = 4)
img_filename <- "rare_plot.png"
## Add if statement to upload image for first time then update if already present
# googledrive::drive_upload(media = tmp_img, type = "png",
#              path =  "defense_presentation",
#              name = img_filename)
googledrive::drive_update(paste0("defense_presentation/", img_filename), tmp_img)
```

