---
title: "Technical Artifacts"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2:
    toc: FALSE
---

<!-- differences between beta diversity metrics  -->

```{r techArtSetup, warning=FALSE, message=FALSE, echo = FALSE}
library(ProjectTemplate)
load.project()
source("src/plot_color_definitions.R")

# library(ggridges)
library(ggpubr)

library(ggthemes)
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

```{r techArtMunge}
## Pre-processing data for figure(s)
qual_df <- seq_qual_comparisons %>%
    mutate(read_dir = paste0(read_dir, "_qual")) %>%
    select(-sd_quality, -cov_quality) %>%
    spread(read_dir, mean_quality) %>%
    separate(replicate, c("biosample_id","t_fctr","seq_run_id"),
             remove = FALSE, extra = "merge")

tech_art_df <- seq_char_comparisons %>%
    select(replicate,  pipe, normalization, metric, cov_num_reads, cov_total_abu) %>%
    left_join(qual_df) %>%
    mutate(f_type = if_else(metric %in% c("unifrac","wunifrac"),
                            "Phylogenetic","Taxonomic"),
           weight = if_else(metric %in% c("unifrac","jaccard"),
                            "Unweighted", "Weighted")) %>%
    mutate(pipe = case_when(pipe == "deblur" ~ "q_deblur",
                            pipe == "qiimeOpenRef" ~ "q_open",
                            pipe == "qiimeClosedRef" ~ "q_closed",
                            pipe == "qiimeDeNovo"~ "q_denovo",
                            TRUE ~ pipe),
           pipe = factor(pipe))

tech_info_df <- tech_art_df %>%
    filter(normalization == "RAW") %>%
    select(biosample_id, t_fctr, seq_run_id, pipe, cov_total_abu, R1_qual, R2_qual) %>%
    distinct()


norm_impact <- tech_art_df %>%
    select(biosample_id, t_fctr, seq_run_id, pipe, normalization, metric, mean_dist) %>%
    spread(normalization, mean_dist) %>%
    gather(normalization, mean_dist, -biosample_id, -t_fctr, -seq_run_id, -pipe, -metric, -RAW) %>%
    mutate(dist_change = mean_dist - RAW) %>%
    filter(!is.na(dist_change)) %>%
    group_by(metric, pipe, normalization, seq_run_id) %>%
    summarise(dist_change = median(dist_change))

```

```{r techMeanDist, fig.cap = "Distribution of mean pairwise beta diversity for PCR replicates by sequencing run and pipeline for raw count data.", fig.height=4, fig.width=6}
tech_art_df %>% filter(normalization == "RAW") %>%
    ggplot() +
    geom_boxplot(aes(x = pipe, y = mean_dist, fill = seq_run_id),
                 outlier.size = 0.5) +
    facet_grid(f_type~weight) + theme_bw() +
    theme(axis.text.x = element_text(angle = -45, hjust = 0),
          legend.position = "bottom") +
    labs(x = "Bioinformatic Pipeline",
         y = "PCR Replicate Beta Diversity",
         fill = "Sequncing Run") +
    scale_fill_manual(values = run_colors)
```

```{r techNormImpactWeighted, fig.cap = "Impact of normalization method on mean weighted (A) and unweighted (B) beta diversity between pcr replicates, for sequencing run with higher quality and total abundance variability, NIST run 1. Data are presented as minimal-ink boxplots, where points indicate median value, gap between point and lines the interquartile range, and lines the boxplot whiskers. Solid black lines represent median value and dashed lines indicate the first and third quartiles of the raw (un-normalized) mean pairwise distances between pcr replicates", fig.height=8, fig.width=7}
norm_level_ord <- c("RAW","rare2000", "rare5000", "rare10000", "rareq15",
                    "CSS", "RLE", "TMM", "TSS", "UQ")

tech_art_df_for_plot_RAW <- tech_art_df %>%
    ungroup() %>%
    filter(normalization == "RAW" & seq_run_id == "nist_run1") %>%
    group_by(pipe,metric,f_type,weight) %>%
    summarize(RAW_median = median(mean_dist),
              RAW_first_quartile = quantile(mean_dist, probs = 0.25),
              RAW_third_quartile = quantile(mean_dist, probs = 0.75))

tech_art_df_for_plot <- tech_art_df %>% ungroup() %>%
    filter(normalization != "RAW", seq_run_id == "nist_run1")

tech_art_df_for_plot <- merge(tech_art_df_for_plot, tech_art_df_for_plot_RAW,
                            by = c("pipe","metric","f_type","weight"))

weighted <- tech_art_df_for_plot %>%
    filter(weight == "Weighted") %>%
    mutate(normalization = factor(normalization,
                                  levels = norm_level_ord, ordered = T)) %>%
    filter(seq_run_id %in% c("nist_run1")) %>%
     ggplot(aes(x = normalization)) +
        geom_blank() +
        geom_ribbon(aes(x = as.numeric(normalization) - 1,
                        ymin = RAW_first_quartile,
                        ymax = RAW_third_quartile),
                    fill = "grey90") +
        geom_line(aes(x = normalization, y = RAW_median,
                  group = interaction(pipe, f_type)),
                  color = "white") +
    # geom_line(aes(x = normalization, y = RAW_first_quartile,
    #               group = interaction(pipe, f_type)),
    #           linetype = "dashed", color = "grey60") +
    # geom_line(aes(x = normalization, y = RAW_third_quartile,
    #               group = interaction(pipe, f_type)),
    #           linetype = "dashed", color = "grey60") +
    geom_tufteboxplot(aes(x = normalization, y = mean_dist, color = normalization,
                          group = interaction(normalization,pipe)),
                      position = position_dodge(), size = 1) +
    facet_grid(f_type~pipe,scales = "free", space = "free") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -45, hjust = 0)) +
    labs(x = "Normalization",
         y = "PCR Replicate Beta Diversity",
         color = "Bioinformatic Pipeline") +
    scale_color_manual(values = norm_colors)

unweighted <- tech_art_df_for_plot %>%
    filter(weight == "Unweighted") %>%
    mutate(normalization = factor(normalization,
                                  levels = norm_level_ord, ordered = T)) %>%
    filter(seq_run_id %in% c("nist_run1")) %>%
    ggplot(aes(x = normalization)) +
        geom_blank() +
        geom_ribbon(aes(x = as.numeric(normalization) - 1,
                        ymin = RAW_first_quartile,
                        ymax = RAW_third_quartile),
                    fill = "grey90") +
        geom_line(aes(x = normalization, y = RAW_median,
                  group = interaction(pipe, f_type)), color = "white") +
    # geom_line(aes(x = normalization, y = RAW_first_quartile,
    #               group = interaction(pipe, f_type)),
    #           linetype = "dashed", color = "grey60") +
    # geom_line(aes(x = normalization, y = RAW_third_quartile,
    #               group = interaction(pipe, f_type)),
    #           linetype = "dashed", color = "grey60") +
    geom_tufteboxplot(aes(x = normalization, y = mean_dist, color = normalization, shape = metric,
                          group = interaction(normalization,pipe)),
                      position = position_dodge(), size = 1) +
    facet_grid(f_type~pipe,scales = "free", space = "free") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -45, hjust = 0)) +
    labs(x = "Normalization",
         y = "PCR Replicate Beta Diversity",
         color = "Bioinformatic Pipeline") +
    scale_color_manual(values = norm_colors)

ggarrange(weighted, unweighted,
          ncol = 1, nrow = 2, labels = "AUTO",
          common.legend = TRUE)
```


We evaluated differences in beta diversity between PCR replicates across sequncing runs to assess bioinformatic pipeline and normalization methods robustness to low quality sequence data and variability in sample total feature abundance.
Higher pairwise distances for NIST run 1 indicates diversity metrics were negatively impacted by larger variation in total abundance across PCR replicates (Fig. \@ref(fig:seqCharPlot)A).
Whereas, higher pairwise distances for JHU run 1 indicates bioinformatic pipelines are negatively impacted by sequencing errors (Fig. \@ref(fig:seqCharPlot)B & C).
The NIST and JHU second runs had low sequencing error rates sample total abundance variation and therefore we would expect the beta diversity between PCR replicates for these runs to be lower than the first runs.


We assessed the impact of sequence quality on the beta diversity between PCR replicates for the different pipelines using the raw count data.
Mean beta diversity between PCR replicates was consistent across the JHU runs for Mothur and DADA2, suggesting they more robust to sequencing errors than the other pipelines evaluated in this study (Fig. \@ref(fig:techMeanDist)).
Conversely, the Deblur pipeline had the highest number of failed samples for the first JHU run, and therefore less robust to sequencing errors compared to the other pipelines (Table \@ref(tab:pipeCharTbl)).
Based on the sequencing run error rates and variation in total abundance we expected the second NIST run to have comparable, if not lower, beta diversity between PCR replicates.
However, the beta diversity was consistently higher for both the NIST runs than the second JHU run and the second NIST run values were closer to the first NIST run than the second JHU run.
The beta diversity between PCR replicates varied by diversity metric (Fig. \@ref(fig:techMeanDist)).
Beta diversity was consistently higher for unweighted compared to weighted metrics and phylogenetic diversity metrics were lower than taxonomic metrics.
Similar pattern in beta diversity for the different sequencing runs and pipelines was observed for the two unweighted beta diversity metrics.


Normalization methods varied in their ability to beta diversity repeatability.
Using data from the fist NIST sequencing run we evaluated the normalization method effect we compared normalized count beta diversity values to normalized count beta diversity (Fig. \@ref(fig:seqCharPlot)).
Most normalization methods had minimal impact on beta diversity repeatability.
However, a number of pipelines had significantly lower weighted beta diversity between PCR replicates when normalized with TMM and RLE compared to the unnormalized count data (Fig. \@ref(fig:techNormImpactWeighted)).
In general beta diversity for the unnormalized counts compared to the normalized counts not significantly different, aside from a few pipeline metric combinations.
The lack of consistent effect on weighted beta diversity metrics for the other normalization methods indicates that there is either no effect or that normalization method effect is metric and pipeline specific.

Performance varied for count data normalized using rarefaction.
Rarefying counts to the total abundance of the 85th most abundance sample (rareq15) resulted in no differences or lowered beta diversity repeatability for PCR replicates for all but the mothur-jaccard and _de novo_-Unifrac pipeline-diversity metric combinations.
Rarefying counts to 2000, improved results for all but closed ref, though the effect was not significant for deblur.
Rarefying to 5000 total abundance yieled similar results to 2000, though the effect was statistically significant for fewer pipeline-metric combinations.
Overall, count data normalized using TMM and RLE had the lowest beta diversity between PCR replicates for weighted metrics, followed by TSS.
For unweighted metrics, rarefying count data to 2000 total abundance resulted in the lowest beta diversity between PCR replicates.

<!-- For discussion
However, the beta diversity was consistently higher for both the NIST runs than the second JHU run and the second NIST run values were closer to the first NIST run than the second JHU run.
    * Potentially due to higher R2 seq quality…
    * Unaccounted for lab effect????
The beta diversity between PCR replicates also varied by diversity metric.
Beta diversity was consistently higher for unweighted compared to weighted metrics.
  * presence - absence
  * Spurious features greater impact on presence absence
  * Unweighted diversity metrics provide information about the low abundant community members.
    - Need presence-absence metrics that are robust to spurious features
-->
