---
title: "PCR Replicate Beta Diversity Repeatability"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2:
    toc: FALSE
---

<!-- differences between beta diversity metrics  -->

```{r techArtSetup, warning=FALSE, message=FALSE, echo = FALSE}
library(ProjectTemplate)
load.project()
source("src/plot_color_definitions.R")

# library(ggridges)
library(ggpubr)

library(ggthemes)
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

```{r techArtMunge}
## Pre-processing data for figure(s)
qual_df <- seq_qual_comparisons %>%
    mutate(read_dir = paste0(read_dir, "_qual")) %>%
    select(-sd_quality, -cov_quality) %>%
    spread(read_dir, mean_quality) %>%
    separate(replicate, c("biosample_id","t_fctr","seq_run_id"),
             remove = FALSE, extra = "merge")

tech_art_df <- seq_char_comparisons %>%
    select(replicate,  pipe, normalization, metric, cov_num_reads, cov_total_abu) %>%
    left_join(qual_df) %>%
    mutate(f_type = if_else(metric %in% c("unifrac","wunifrac"),
                            "Phylogenetic","Taxonomic"),
           weight = if_else(metric %in% c("unifrac","jaccard"),
                            "Unweighted", "Weighted")) %>%
    mutate(pipe = case_when(pipe == "deblur" ~ "q_deblur",
                            pipe == "qiimeOpenRef" ~ "q_open",
                            pipe == "qiimeClosedRef" ~ "q_closed",
                            pipe == "qiimeDeNovo"~ "q_denovo",
                            TRUE ~ pipe),
           pipe = factor(pipe))

tech_info_df <- tech_art_df %>%
    filter(normalization == "RAW") %>%
    select(biosample_id, t_fctr, seq_run_id, pipe, cov_total_abu, R1_qual, R2_qual) %>%
    distinct()


norm_impact <- tech_art_df %>%
    select(biosample_id, t_fctr, seq_run_id, pipe, normalization, metric, mean_dist) %>%
    spread(normalization, mean_dist) %>%
    gather(normalization, mean_dist, -biosample_id, -t_fctr, -seq_run_id, -pipe, -metric, -RAW) %>%
    mutate(dist_change = mean_dist - RAW) %>%
    filter(!is.na(dist_change)) %>%
    group_by(metric, pipe, normalization, seq_run_id) %>%
    summarise(dist_change = median(dist_change))

```

```{r techMeanDist, fig.cap = "Distribution of mean pairwise PCR replicate beta-diversity for by sequencing run and pipeline for un-normalized count data.", fig.height=4, fig.width=6}
tech_art_df %>% filter(normalization == "RAW") %>%
    ggplot() +
    geom_boxplot(aes(x = pipe, y = mean_dist, fill = seq_run_id),
                 outlier.size = 0.5) +
    facet_grid(f_type~weight) + theme_bw() +
    theme(axis.text.x = element_text(angle = -45, hjust = 0),
          legend.position = "bottom") +
    labs(x = "Bioinformatic Pipeline",
         y = "PCR Replicate Beta-Diversity",
         fill = "Sequncing Run") +
    scale_fill_manual(values = run_colors)
```

```{r techNormImpactWeighted, fig.cap = "Impact of normalization method on mean weighted (A) and unweighted (B) pcr replicates beta-diversity, for sequencing run with higher quality and total abundance variability, NIST run 1. Data are presented as minimal-ink boxplots, where points indicate median value, gap between point and lines the interquartile range, and lines the boxplot whiskers. Solid black lines represent median value and dashed lines indicate the first and third quartiles of the raw (un-normalized) mean pairwise distances between pcr replicates.", fig.height=8, fig.width=7}
norm_level_ord <- c("RAW","rare2000", "rare5000", "rare10000", "rareq15",
                    "CSS", "RLE", "TMM", "TSS", "UQ")

tech_art_df_for_plot_RAW <- tech_art_df %>%
    ungroup() %>%
    filter(normalization == "RAW" & seq_run_id == "nist_run1") %>%
    group_by(pipe,metric,f_type,weight) %>%
    summarize(RAW_median = median(mean_dist),
              RAW_first_quartile = quantile(mean_dist, probs = 0.25),
              RAW_third_quartile = quantile(mean_dist, probs = 0.75))

tech_art_df_for_plot <- tech_art_df %>% ungroup() %>%
    filter(normalization != "RAW", seq_run_id == "nist_run1")

tech_art_df_for_plot <- merge(tech_art_df_for_plot, tech_art_df_for_plot_RAW,
                            by = c("pipe","metric","f_type","weight"))

weighted <- tech_art_df_for_plot %>%
    filter(weight == "Weighted") %>%
    mutate(normalization = factor(normalization,
                                  levels = norm_level_ord, ordered = T)) %>%
    filter(seq_run_id %in% c("nist_run1")) %>%
     ggplot(aes(x = normalization)) +
        geom_blank() +
        geom_ribbon(aes(x = as.numeric(normalization) - 1,
                        ymin = RAW_first_quartile,
                        ymax = RAW_third_quartile),
                    fill = "grey90") +
        geom_line(aes(x = normalization, y = RAW_median,
                  group = interaction(pipe, f_type)),
                  color = "white") +
    # geom_line(aes(x = normalization, y = RAW_first_quartile,
    #               group = interaction(pipe, f_type)),
    #           linetype = "dashed", color = "grey60") +
    # geom_line(aes(x = normalization, y = RAW_third_quartile,
    #               group = interaction(pipe, f_type)),
    #           linetype = "dashed", color = "grey60") +
    geom_tufteboxplot(aes(x = normalization, y = mean_dist, color = normalization,
                          group = interaction(normalization,pipe)),
                      position = position_dodge(), size = 1) +
    facet_grid(f_type~pipe,scales = "free", space = "free") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -45, hjust = 0),
          legend.position = "none") +
    labs(x = "Normalization",
         y = "PCR Replicate Beta-Diversity",
         color = "Bioinformatic Pipeline") +
    scale_color_manual(values = norm_colors)

unweighted <- tech_art_df_for_plot %>%
    filter(weight == "Unweighted") %>%
    mutate(normalization = factor(normalization,
                                  levels = norm_level_ord, ordered = T)) %>%
    filter(seq_run_id %in% c("nist_run1")) %>%
    ggplot(aes(x = normalization)) +
        geom_blank() +
        geom_ribbon(aes(x = as.numeric(normalization) - 1,
                        ymin = RAW_first_quartile,
                        ymax = RAW_third_quartile),
                    fill = "grey90") +
        geom_line(aes(x = normalization, y = RAW_median,
                  group = interaction(pipe, f_type)), color = "white") +
    # geom_line(aes(x = normalization, y = RAW_first_quartile,
    #               group = interaction(pipe, f_type)),
    #           linetype = "dashed", color = "grey60") +
    # geom_line(aes(x = normalization, y = RAW_third_quartile,
    #               group = interaction(pipe, f_type)),
    #           linetype = "dashed", color = "grey60") +
    geom_tufteboxplot(aes(x = normalization, y = mean_dist, color = normalization, shape = metric,
                          group = interaction(normalization,pipe)),
                      position = position_dodge(), size = 1) +
    facet_grid(f_type~pipe,scales = "free", space = "free") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -45, hjust = 0),
          legend.position = "none") +
    labs(x = "Normalization",
         y = "PCR Replicate Beta-Diversity",
         color = "Bioinformatic Pipeline") +
    scale_color_manual(values = norm_colors)

ggarrange(weighted, unweighted,
          ncol = 1, nrow = 2, labels = "AUTO",
          common.legend = TRUE)
```


Next, we evaluated differences in beta-diversity between un-normalized PCR replicates across sequencing runs and pipelines.
PCR replicate beta-diversity varied by diversity metric (Fig. \@ref(fig:seqCharPlot)).
Beta-diversity was consistently higher for unweighted compared to weighted metrics and phylogenetic diversity metrics were lower than taxonomic metrics.
We expected to see higher pairwise distances for the lower quality JHU1 run compared to the higher quality JHU2 run.
This was true for the QIIME clustering pipelines, however the Mothur and DADA2 mean PCR replicate beta diversity was consistent across the JHU runs, suggesting that these pipelines are more robust to sequencing errors (Fig. \@ref(fig:seqCharPlot)).
Conversely, with the highest number of failed samples for the first JHU run, the Deblur pipeline was the least robust to sequencing errors (Table \@ref(tab:pipeCharTbl)).
We also expected that the second NIST run, which had higher read quality and lower total abundance variability, to have comparable, if not lower, PCR replicates beta-diversity. This was not the case.  Beta-diversity was consistently higher for both NIST runs compared to the high quality, low variability second JHU run. Furthermore, the second NIST run values were more similar to the first high quality and high total abundance variability NIST run, than the second JHU run.


Normalization methods varied in their ability to improve beta-diversity repeatability.
When comparing normalized to un-normalized PCR replicate beta-diversity, we observed that most normalization methods had minimal impact on weighted metrics (Fig. \@ref(fig:techNormImpactWeighted)A).
For a number of pipelines, TMM and RLE normalization methods significantly lowered weighted PCR replicate beta-diversity (Fig. \@ref(fig:techNormImpactWeighted)A).
For unweighted metrics (Fig. \@ref(fig:techNormImpactWeighted)B), rarefying count data to 2000 total feature abundance resulted in the lowest beta-diversity between PCR replicates.
While rarefying counts to the total abundance of the 15th most abundant sample (rareq15) resulted in no differences or lower PCR replicate beta-diversity for most pipeline-diversity metric combinations, resulting in higher weighed beta-diversity for QIIME clustering techniques.
Rarefaction to this level is also most susceptible to sample loss and should not be used as it results in unnecessary loss of statistical power (Supplemental Table X).


<!-- For discussion
However, the beta diversity was consistently higher for both the NIST runs than the second JHU run and the second NIST run values were closer to the first NIST run than the second JHU run.
    * Potentially due to higher R2 seq quality…
    * Unaccounted for lab effect????
The beta diversity between PCR replicates also varied by diversity metric.
Beta diversity was consistently higher for unweighted compared to weighted metrics.
  * presence - absence
  * Spurious features greater impact on presence absence
  * Unweighted diversity metrics provide information about the low abundant community members.
    - Need presence-absence metrics that are robust to spurious features
-->
