---
title: "Technical Artifacts"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2: 
    toc: FALSE
---

```{r techArtSetup, warning=FALSE, message=FALSE, echo = FALSE}
library(ProjectTemplate)
load.project()
library(ggridges)
library(ggpubr)
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

```{r techArtMunge}
## Pre-processing data for figure(s)
qual_df <- seq_qual_comparisons %>% 
    mutate(read_dir = paste0(read_dir, "_qual")) %>% 
    select(-sd_quality, -cov_quality) %>% 
    spread(read_dir, mean_quality) %>% 
    separate(replicate, c("biosample_id","t_fctr","seq_run_id"), remove = FALSE,extra = "merge")

tech_art_df <- seq_char_comparisons %>% 
    select(replicate,  pipe, normalization, metric, cov_num_reads, cov_total_abu) %>% 
    left_join(qual_df)

tech_info_df <- tech_art_df %>% 
    filter(normalization == "RAW") %>% 
    select(biosample_id, t_fctr, seq_run_id, pipe, cov_total_abu, R1_qual, R2_qual) %>% 
    distinct() 


norm_impact <- tech_art_df %>% 
    select(biosample_id, t_fctr, seq_run_id, pipe, normalization, metric, mean_dist) %>% 
    spread(normalization, mean_dist) %>% 
    gather(normalization, mean_dist, -biosample_id, -t_fctr, -seq_run_id, -pipe, -metric, -RAW) %>% 
    mutate(dist_change = mean_dist - RAW) %>% 
    filter(!is.na(dist_change)) %>% 
    group_by(metric, pipe, normalization, seq_run_id) %>% 
    summarise(dist_change = median(dist_change)) %>% 
        mutate(f_type = if_else(metric %in% c("unifrac","wunifrac"), 
                            "Phylogenetic","Taxonomic"),
           weight = if_else(metric %in% c("unifrac","jaccard"),
                            "Unweighted", "Weighted")) 
```
  
```{r techTotalAbuCOV, fig.cap = "A. Distribution of the coefficienct of variation of total abundance for PCR replicates by bioinformatic pipeline and sequencing run across B. Mean read quality score for PCR replicates across sequencing runs for forward and reverse reads."}
# ggplot(tech_info_df) + 
#     geom_boxplot(aes(x = seq_run_id, y = cov_total_abu)) + 
#     labs(x = "Sequencing Run", y = "COV(Total Abundance)") + facet_wrap(~pipe) + theme_bw()

tech_info_df_summary<-tech_info_df %>% group_by(seq_run_id, pipe) %>% summarize(mean_cov_total_abu=mean(cov_total_abu))

cov_plot<-ggplot(tech_info_df_summary, aes(x = pipe, y = mean_cov_total_abu, group=seq_run_id, color=seq_run_id)) + 
    geom_point() + geom_line()+
    labs(x = "Sequencing Run", y = "Mean COV(Total Abundance)") + theme_bw()+
    theme(axis.text.x=element_text(angle = 45, hjust = 1))

read_qual_plot<-tech_info_df %>% 
    select(biosample_id, t_fctr, seq_run_id, R1_qual, R2_qual) %>% 
    distinct() %>% 
    gather("read_dir","mean_qual", -biosample_id, -t_fctr, -seq_run_id) %>% 
    ggplot() + 
    geom_boxplot(aes(x = seq_run_id, y = mean_qual, group=interaction(seq_run_id,read_dir), 
                     fill=read_dir), position=position_dodge()) + 
    labs(x = "Sequencing Run", y = "Mean Quality Score", fill = "Read Direction") + 
    theme_bw()+theme(axis.text.x=element_text(angle = 45, hjust = 1))

ggarrange(cov_plot, read_qual_plot)

```

```{r techMeanDist, message = FALSE, fig.cap = "Distribution of mean pairwise beta diversity for PCR replicates by sequencing run and pipeline for raw count data."}
tech_art_df %>% filter(normalization == "RAW") %>% 
    ggplot() + 
    geom_density_ridges(aes(x = mean_dist, y = seq_run_id), 
                         stat = "binline",
                        draw_baseline = FALSE,
                        scale = 1,
                    alpha = 0.5) + 
    facet_grid(pipe~metric, scales = "free_x") + theme_bw()
```

```{r techNormImpactALL, fig.cap = "Impact of normalization method on mean beta diversity between pcr replicates, accross sequencing runs."}
norm_impact %>% 
    ggplot() + 
    geom_hline(aes(yintercept = 0)) + 
    geom_point(aes(x = pipe, y = dist_change, color = seq_run_id, shape=metric), size=2, position=position_jitter(width=0.3)) + 
    facet_wrap(~factor(normalization, levels=c("rare2000", "rare5000", "rare10000", "rareq15", "CSS", "RLE", "TMM", "TSS", "UQ"), ordered = T), ncol=5) + 
    theme_bw() + 
    theme(legend.position = "bottom", axis.text.x = element_text(angle = 90)) + 
    labs(x = "Pipeline", y = "Change in Mean Beta Diversity", color = "Norm") 
```

__Key Points__  

* JHU run 1 had lower sequence quality (boxplots) (Fig. \@ref(fig:techTotalAbuCOV)).  
* NIST run 1 had greater variabiltiy in sample total abundance (lineplot) (Fig. \@ref(fig:techTotalAbuCOV)).  
* Mean pairwise distance was greater for qiime De-novo, open and closed reference pipelines for JHU1 relative to the other sequencing runs for all metrics excluding weighted unifrac (Fig. \@ref(fig:techMeanDist)).   
* Qiime De novo had high pairwise distance across sequencing runs for Unifrac and low for weighted Unifrac.  
* Pairwise distances varies by metric.  
* DADA2 pairwise distances greater for NIST1 and NIST2 compared to JHU runs, JHU2 had the lowest pairwise distance.  
* Mothur and dada, consistent results for JHU runs, better able to account for sequencing errors. 
* Deblur pipeline failed for JHU1.  
* Higher pairwise distances for NIST runs indicate diversity metrics impacted by larger variation in total abundance across PCR replicates.  

__Additional Points__  

* De novo weighted Unifrac low for QIIME De novo: This is potentially due to large number of singletons in weighted unifrac dataset, ~120K out of ~180K total features. These singletons are likely sequencing errors and therefore closely related to other taxa therefore minimally impact the weighted unifrac results.  