---
title: "Technical Artifacts"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2: 
    toc: FALSE
---

<!-- differences between beta diversity metrics  -->

```{r techArtSetup, warning=FALSE, message=FALSE, echo = FALSE}
library(ProjectTemplate)
load.project()
source("src/plot_color_definitions.R")

# library(ggridges)
library(ggpubr)

library(ggthemes)
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

```{r techArtMunge}
## Pre-processing data for figure(s)
qual_df <- seq_qual_comparisons %>% 
    mutate(read_dir = paste0(read_dir, "_qual")) %>% 
    select(-sd_quality, -cov_quality) %>% 
    spread(read_dir, mean_quality) %>% 
    separate(replicate, c("biosample_id","t_fctr","seq_run_id"), 
             remove = FALSE, extra = "merge")

tech_art_df <- seq_char_comparisons %>% 
    select(replicate,  pipe, normalization, metric, cov_num_reads, cov_total_abu) %>% 
    left_join(qual_df) %>% 
    mutate(f_type = if_else(metric %in% c("unifrac","wunifrac"), 
                            "Phylogenetic","Taxonomic"),
           weight = if_else(metric %in% c("unifrac","jaccard"),
                            "Unweighted", "Weighted")) 

tech_info_df <- tech_art_df %>% 
    filter(normalization == "RAW") %>% 
    select(biosample_id, t_fctr, seq_run_id, pipe, cov_total_abu, R1_qual, R2_qual) %>% 
    distinct() 


norm_impact <- tech_art_df %>% 
    select(biosample_id, t_fctr, seq_run_id, pipe, normalization, metric, mean_dist) %>% 
    spread(normalization, mean_dist) %>% 
    gather(normalization, mean_dist, -biosample_id, -t_fctr, -seq_run_id, -pipe, -metric, -RAW) %>% 
    mutate(dist_change = mean_dist - RAW) %>% 
    filter(!is.na(dist_change)) %>% 
    group_by(metric, pipe, normalization, seq_run_id) %>% 
    summarise(dist_change = median(dist_change)) 
        
```

```{r techMeanDist, fig.cap = "Distribution of mean pairwise beta diversity for PCR replicates by sequencing run and pipeline for raw count data.", fig.height=4, fig.width=6}
tech_art_df %>% mutate(pipe = if_else(pipe == "deblur", "q_deblur", 
                          if_else(pipe == "qiimeOpenRef", "q_open", 
                                  if_else(pipe == "qiimeClosedRef", "q_closed",
                                          if_else(pipe == "qiimeDeNovo", "q_denovo",
                                          pipe)))),
           pipe = factor(pipe)) %>% 
    filter(normalization == "RAW") %>% 
    ggplot() + 
    # geom_violin(aes(x = pipe, y = mean_dist), fill = "grey60") +
    geom_boxplot(aes(x = pipe, y = mean_dist, fill = seq_run_id),
                 outlier.size = 0.5) + 
    facet_grid(f_type~weight) + theme_bw() + 
    theme(axis.text.x = element_text(angle = -45, hjust = 0),
          legend.position = "bottom") + 
    labs(x = "Bioinformatic Pipeline", 
         y = "PCR Replicate Beta Diversity", 
         fill = "Sequncing Run") + 
    scale_fill_manual(values = run_colors)
```

```{r techNormImpactWeighted, fig.cap = "Impact of normalization method on mean weighted (A) and unweighted (B) beta diversity between pcr replicates, for sequencing run with higher quality and total abundance variability, NIST run 1. Data are presented as minimal-ink boxplots, where points indicate median value, gap between point and lines the interquartile range, and lines the boxplot whiskers. Solid black lines represent median value and dashed lines indicate the first and third quartiles of the raw (un-normalized) mean pairwise distances between pcr replicates", fig.height=8, fig.width=7}
norm_level_ord <- c("RAW","rare2000", "rare5000", "rare10000", "rareq15", 
                    "CSS", "RLE", "TMM", "TSS", "UQ") 

tech_art_df_for_plot_RAW <- tech_art_df %>% 
    ungroup() %>% 
    filter(normalization == "RAW" & seq_run_id == "nist_run1") %>% 
    group_by(pipe,metric,f_type,weight) %>%
    summarize(RAW_median = median(mean_dist),
              RAW_first_quartile = quantile(mean_dist, probs = 0.25),
              RAW_third_quartile = quantile(mean_dist, probs = 0.75))

tech_art_df_for_plot <- tech_art_df %>% ungroup() %>% 
    filter(normalization != "RAW", seq_run_id == "nist_run1")

tech_art_df_for_plot <- merge(tech_art_df_for_plot, tech_art_df_for_plot_RAW, 
                            by = c("pipe","metric","f_type","weight"))

weighted_tech_art_norm_eff <- tech_art_df_for_plot %>% 
    mutate(pipe = if_else(pipe == "deblur", "q_deblur", 
                          if_else(pipe == "qiimeOpenRef", "q_open", 
                                  if_else(pipe == "qiimeClosedRef", "q_closed",
                                          if_else(pipe == "qiimeDeNovo", "q_denovo",
                                          pipe)))),
           pipe = factor(pipe)) %>%
    filter(weight == "Weighted") %>% 
    # filter(!(normalization %in% c("RLE","rare2000", "rare10000", "UQ"))) %>% 
    mutate(normalization = factor(normalization, 
                       levels = norm_level_ord, ordered = T)) %>% 
    filter(seq_run_id %in% c("nist_run1")) %>% 
     ggplot(aes(x = normalization)) +
        geom_blank() + 
        geom_ribbon(aes(x = as.numeric(normalization) - 1, 
                        ymin = RAW_first_quartile, 
                        ymax = RAW_third_quartile),
                    fill = "grey90") + 
        geom_line(aes(x = normalization, y = RAW_median, 
                  group = interaction(pipe, f_type)),
                  color = "white") +
    # geom_line(aes(x = normalization, y = RAW_first_quartile, 
    #               group = interaction(pipe, f_type)),
    #           linetype = "dashed", color = "grey60") +
    # geom_line(aes(x = normalization, y = RAW_third_quartile, 
    #               group = interaction(pipe, f_type)), 
    #           linetype = "dashed", color = "grey60") +
    geom_tufteboxplot(aes(x = normalization, y = mean_dist, color = normalization, 
                          group = interaction(normalization,pipe)), 
                      position = position_dodge(), size = 1) + 
    facet_grid(f_type~pipe,scales = "free", space = "free") + 
    theme_bw() + 
    theme(axis.text.x = element_text(angle = -45, hjust = 0), 
          legend.position = "none") + 
    labs(x = "Normalization", y = "Mean Beta Diversity", color = "Pipeline") + 
    scale_color_manual(values = norm_colors)  

unweighted_tech_art_norm_eff <- tech_art_df_for_plot %>% 
    filter(weight == "Unweighted") %>%
    mutate(normalization = factor(normalization, 
                       levels = norm_level_ord, ordered = T)) %>% 
    filter(seq_run_id %in% c("nist_run1")) %>% 
    ggplot(aes(x = normalization)) +
        geom_blank() + 
        geom_ribbon(aes(x = as.numeric(normalization) - 1, 
                        ymin = RAW_first_quartile, 
                        ymax = RAW_third_quartile),
                    fill = "grey90") + 
        geom_line(aes(x = normalization, y = RAW_median, 
                  group = interaction(pipe, f_type)), color = "white") +
    # geom_line(aes(x = normalization, y = RAW_first_quartile, 
    #               group = interaction(pipe, f_type)),
    #           linetype = "dashed", color = "grey60") +
    # geom_line(aes(x = normalization, y = RAW_third_quartile, 
    #               group = interaction(pipe, f_type)), 
    #           linetype = "dashed", color = "grey60") +
    geom_tufteboxplot(aes(x = normalization, y = mean_dist, color = normalization, 
                          group = interaction(normalization,pipe)), 
                      position = position_dodge(), size = 1) + 
    facet_grid(f_type~pipe,scales = "free", space = "free") + 
    theme_bw() + 
    theme(axis.text.x = element_text(angle = -45, hjust = 0), 
          legend.position = "none") + 
    labs(x = "Normalization", y = "Mean Beta Diversity", color = "Pipeline") + 
    scale_color_manual(values = norm_colors) 

ggarrange(weighted_tech_art_norm_eff, unweighted_tech_art_norm_eff,
          ncol = 1, nrow = 2, labels = "AUTO")
```


We evaluated differences in beta diversity between PCR replicates across sequncing runs to assess how robust the different bioinformatic pipelines and normalization methods are to low quality sequence data and variability in per sample total abundance. 
Higher pairwise distances for NIST runs indicate diversity metrics negatively impacted by larger variation in total abundance across PCR replicates (Fig. \@ref(fig:seqCharPlot)A). 
Whereas, higher pairwise distances for JHU run 1 indicates bioinformatic pipelines are less robust to sequencing errors (Fig. \@ref(fig:seqCharPlot)B & C). A linear model was used to test for significant differences in PCR replicates beta diversity by pipeline and sequencing run for raw count data. 
Pipeline, sequencing run, and interaction between the two were significant for the four distance metrics (Supplemental Results??). 
Tukey honest significant differences was used to test for significant pairwise differences (Supplemental Results??). 

<!-- Not sure about reporting statistical significance, what multiple comparison correction should I use  -->
Nearly all pipeline comparisons were significantly different except for for Bray Curtis (unweighted taxonomic), where the QIIME _de novo_, closed-reference, and open-reference were not significantly different from eachother, as well as mothur and DADA2 were not significantly different from eachother. 
<!-- Overall Mothur, dada, and deblur had lower mean pairwise distances compared to the QIIME open-reference, closed-reference, and _de novo_ results were other pipelines for three of the four diversity metrics.  -->
<!-- For weighted UniFrac QIIME _de novo_ mean pairwise distances was similar to the Mothur, DADA2, and Deblur results.  -->
<!-- The low weighted UniFrac for QIIME _de novo_ is potentially due to large number of singletons in the dataset, ~120K out of ~180K total features (Table \@ref(tab:pipeCharTbl)).  -->
<!-- These singletons are likely sequencing errors and therefore closely related to other taxa therefore minimally impact the weighted unifrac results.   -->
DADA2 pairwise distances were significantly greater for NIST1 and NIST2 (which had greater variability in library size) compared to the JHU2 run which had the lowest pairwise distance.  
Mean pairwise distances were consistent across the JHU runs for Mothur and DADA2, suggesting they are better able to account for sequencing errors than other pipelines evaluated in this study. 
Conversely, the Deblur pipeline had the highest number of failed samples for JHU1, suggesting it is less robust to sequencing errors compared to the other pipelines (Table \@ref(tab:pipeCharTbl)).   

* Across pipelines NIST 1 and NIST 2 greater PCR replicate difference  
* QIIME pipelines less robust to lower seq quality  
* Expected lower diff for NIST 2 compared to JHU2 but consistent results with NIST 1, not sure why. Potentially due to higher R2 seq quality…  
* Unaccounted for lab effect????  
* Deblur failed for JHU1  
* Lack of difference in pcr dist for JHU1 and JHU2 with mothur and dada, similar repeat between low and high quality runs????  




__Impact of normalization methods on technical artifacts.__ 

- Compared results across normalization methods for NIST run 1, low error rate, high total abundance COV (Fig. \@ref(fig:seqCharPlot)).   
- Methods had minimal impact, TMM and RLE tended to lower pairwise dist, rare15 increased (Fig. \@ref(fig:techNormImpactWeighted))  
* CSS no significant difference for Bray, Weighted UniFrac significant decrease for deblur and open ref but higher for DADA
* No difference for UQ  
* TSS lower for Bray excluding deblur and dada, lower for Open ref and bray for Wunifrac  
* Rarefaction  
- Rareq15 no differences or lowered repeat for all but Jaccard mothur and Denovo Unifrac  
- Norm 2000  
    - improved repeat for all but closed ref (unweighted), not significant for deblur  
    - weighted imporved for mothur bray and deblur, open ref wunif, not significantly difference for rest  
- Norm 5000 similar results to 2000 but fewer significant.  

interpretation  
* Norm TMM and RLE best for weighted metrics, followed by TSS  
* Mixed results for rarefaction  
* rare2000 best for unweighted  