---
title: "Technical Artifacts"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2: 
    toc: FALSE
---

<!-- differences between beta diversity metrics  -->

```{r techArtSetup, warning=FALSE, message=FALSE, echo = FALSE}
library(ProjectTemplate)
load.project()

library(ggthemes)
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

```{r techArtMunge}
## Pre-processing data for figure(s)
qual_df <- seq_qual_comparisons %>% 
    mutate(read_dir = paste0(read_dir, "_qual")) %>% 
    select(-sd_quality, -cov_quality) %>% 
    spread(read_dir, mean_quality) %>% 
    separate(replicate, c("biosample_id","t_fctr","seq_run_id"), remove = FALSE,extra = "merge")

tech_art_df <- seq_char_comparisons %>% 
    select(replicate,  pipe, normalization, metric, cov_num_reads, cov_total_abu) %>% 
    left_join(qual_df) %>% 
    mutate(f_type = if_else(metric %in% c("unifrac","wunifrac"), 
                            "Phylogenetic","Taxonomic"),
           weight = if_else(metric %in% c("unifrac","jaccard"),
                            "Unweighted", "Weighted")) 

tech_info_df <- tech_art_df %>% 
    filter(normalization == "RAW") %>% 
    select(biosample_id, t_fctr, seq_run_id, pipe, cov_total_abu, R1_qual, R2_qual) %>% 
    distinct() 


norm_impact <- tech_art_df %>% 
    select(biosample_id, t_fctr, seq_run_id, pipe, normalization, metric, mean_dist) %>% 
    spread(normalization, mean_dist) %>% 
    gather(normalization, mean_dist, -biosample_id, -t_fctr, -seq_run_id, -pipe, -metric, -RAW) %>% 
    mutate(dist_change = mean_dist - RAW) %>% 
    filter(!is.na(dist_change)) %>% 
    group_by(metric, pipe, normalization, seq_run_id) %>% 
    summarise(dist_change = median(dist_change)) 
        
```
  
```{r techTotalAbuCOV, fig.cap = "A. Distribution of the coefficienct of variation of total abundance for PCR replicates by bioinformatic pipeline and sequencing run across B. Mean read quality score for PCR replicates across sequencing runs for forward and reverse reads.", eval = FALSE}

cov_plot<-ggplot(tech_info_df) +
    geom_boxplot(aes(x = reorder(interaction(pipe, seq_run_id), cov_total_abu, FUN=median), 
                     y = cov_total_abu, fill=seq_run_id)) +
    labs(x = "Pipeline", y = "COV(Total Abundance)") + 
    theme_bw() + 
    coord_flip() +
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) +
    scale_fill_manual(values=run_colors)

#tech_info_df_summary<-tech_info_df %>% group_by(seq_run_id, pipe) %>% summarize(mean_cov_total_abu=mean(cov_total_abu))
# 
# cov_plot<-ggplot(tech_info_df_summary, aes(x = pipe, y = mean_cov_total_abu, group=seq_run_id, color=seq_run_id)) + 
#     geom_point() + geom_line()+
#     labs(x = "Sequencing Run", y = "Mean COV(Total Abundance)") + theme_bw()+
#     theme(axis.text.x=element_text(angle = 45, hjust = 1))

read_qual_plot<-tech_info_df %>% 
    select(biosample_id, t_fctr, seq_run_id, R1_qual, R2_qual) %>% 
    distinct() %>% 
    gather("read_dir","mean_qual", -biosample_id, -t_fctr, -seq_run_id) %>% 
    ggplot() + 
    geom_boxplot(aes(x = seq_run_id, y = mean_qual, 
                     group=interaction(seq_run_id,read_dir), fill = read_dir), 
                 position=position_dodge()) + 
    labs(x = "Sequencing Run", y = "Mean Quality Score", fill = "Read Direction") + 
    theme_bw()+ 
    theme(axis.text.x=element_text(angle = 45, hjust = 1)) + 
    scale_fill_manual(values=c("black", "darkgray"))

ggarrange(cov_plot, read_qual_plot)

```

```{r techMeanDist_version1, message = FALSE, fig.cap = "Distribution of mean pairwise beta diversity for PCR replicates by sequencing run and pipeline for raw count data.", eval = FALSE}
tech_art_df %>% filter(normalization == "RAW") %>% 
    ggplot() + 
    geom_density_ridges(aes(x = mean_dist, y = seq_run_id), 
                         stat = "binline",
                        draw_baseline = FALSE,
                        scale = 1,
                    alpha = 0.5) + 
    facet_grid(pipe~metric, scales = "free_x") + theme_bw()
```

```{r techMeanDist, fig.cap = "Distribution of mean pairwise beta diversity for PCR replicates by sequencing run and pipeline for raw count data."}
tech_art_df %>% filter(normalization == "RAW") %>% 
    ggplot() + 
    geom_boxplot(aes(x = pipe, y = mean_dist, fill = seq_run_id),
                 outlier.size = 0.5) + 
    facet_grid(f_type~weight) + theme_bw() + 
    theme(axis.text.x = element_text(angle = -45, hjust = 0),
          legend.position = "bottom") + 
    labs(x = "Bioinformatic Pipeline", 
         y = "Mean Distance", 
         fill = "Sequncing Run") + 
    scale_fill_manual(values = run_colors)
```

```{r techNormImpactALL, fig.cap = "Impact of normalization method on mean weighted beta diversity between pcr replicates, for sequencing run with higher quality and total abundance variability, NIST run 1. Data are presented as minimal-ink boxplots, where points indicate median value, gap between point and lines the interquartile range, and lines the boxplot whiskers. Solid black lines represent median value and dashed lines indicate the first and third quartiles of the raw (un-normalized) mean pairwise distances between pcr replicates", fig.height=6, fig.width=12}
norm_level_ord <- c("RAW","rare2000", "rare5000", "rare10000", "rareq15", 
                    "CSS", "RLE", "TMM", "TSS", "UQ") 
tech_art_df_for_plot_RAW <- tech_art_df %>% 
    ungroup() %>% 
    filter(normalization=="RAW" & seq_run_id=="nist_run1") %>% 
    group_by(pipe,metric,f_type,weight) %>%
    summarize(RAW_median= as.numeric(as.character(summary(mean_dist)[3])),
              RAW_first_quartile=as.numeric(as.character(summary(mean_dist)[2])),
              RAW_third_quartile=as.numeric(as.character(summary(mean_dist)[5])))

tech_art_df_for_plot <- tech_art_df %>% ungroup() %>% filter(normalization!="RAW", seq_run_id=="nist_run1")
tech_art_df_for_plot<-merge(tech_art_df_for_plot, tech_art_df_for_plot_RAW, 
                            by=c("pipe","metric","f_type","weight"))

tech_art_df_for_plot %>% 
    filter(weight=="Weighted") %>%
    mutate(normalization = factor(normalization, 
                       levels = norm_level_ord, ordered = T)) %>% 
    filter(seq_run_id %in% c("nist_run1")) %>% 
    ggplot() + 
    geom_tufteboxplot(aes(x = normalization, y = mean_dist, color = normalization, 
                          group=interaction(normalization,pipe)), position = position_dodge(), size=1) + 
    geom_line(aes(x=normalization, y=RAW_median, group=interaction(pipe, f_type)))+
    geom_line(aes(x=normalization, y=RAW_first_quartile, group=interaction(pipe, f_type)),linetype="dashed")+
    geom_line(aes(x=normalization, y=RAW_third_quartile, group=interaction(pipe, f_type)), linetype="dashed")+
    facet_grid(f_type~pipe,scales = "free", space = "free") + 
    theme_bw() + 
    theme(axis.text.x = element_text(angle = -45, hjust = 0), 
          legend.position = "none") + 
    labs(x = "Normalization", 
         y = "Change in Mean Beta Diversity", 
         color = "Pipeline") + 
    scale_color_manual(values = norm_colors)
```


```{r techNormImpactALL2, fig.cap = "Impact of normalization method on mean unweighted beta diversity between pcr replicates, for sequencing run with higher quality and total abundance variability, NIST run 1. Data are presented as minimal-ink boxplots, where points indicate median value, gap between point and lines the interquartile range, and lines the boxplot whiskers. Solid black lines represent median value and dashed lines indicate the first and third quartiles of the raw (un-normalized) mean pairwise distances between pcr replicates", fig.height=6, fig.width=12, eval=F}

tech_art_df_for_plot %>% 
    filter(weight=="Unweighted") %>%
    mutate(normalization = factor(normalization, 
                       levels = norm_level_ord, ordered = T)) %>% 
    filter(seq_run_id %in% c("nist_run1")) %>% 
    ggplot() + 
    geom_tufteboxplot(aes(x = normalization, y = mean_dist, color = normalization, 
                          group=interaction(normalization,pipe)), position = position_dodge(), size=1) + 
    geom_line(aes(x=normalization, y=RAW_median, group=interaction(pipe, f_type)))+
    geom_line(aes(x=normalization, y=RAW_first_quartile, group=interaction(pipe, f_type)),linetype="dashed")+
    geom_line(aes(x=normalization, y=RAW_third_quartile, group=interaction(pipe, f_type)), linetype="dashed")+
    facet_grid(f_type~pipe,scales = "free", space = "free") + 
    theme_bw() + 
    theme(axis.text.x = element_text(angle = -45, hjust = 0), legend.position = "none") + 
    labs(x = "Normalization", y = "Change in Mean Beta Diversity", 
         color = "Pipeline") +scale_color_manual(values=norm_colors)
```
We evaluated differences in beta diversity between PCR replicates between the four sequncing runs to assess how robust the different bioinformatic pipelines and normalization methods are to low quality sequence data and variability in per sample total abundance. 
Higher pairwise distances for NIST runs indicate diversity metrics negatively impacted by larger variation in total abundance across PCR replicates. 
Whereas higher pairwise distances for JHU run 1 indicates bioinformatic pipelines are less robust to sequencing errors. 
Mean pairwise distance was greater for qiime De-novo, open and closed reference pipelines for JHU1 (lower quality and greater variability) relative to the other sequencing runs for all metrics excluding weighted UniFrac (Fig. \@ref(fig:techMeanDist)).
QIIME _de novo_ had high mean pairwise distance across sequencing runs for unweighted UniFrac and low for weighted Unifrac. 
The low weighted UniFrac for QIIME _de novo_ is potentially due to large number of singletons in the dataset, ~120K out of ~180K total features. 
These singletons are likely sequencing errors and therefore closely related to other taxa therefore minimally impact the weighted unifrac results.  
DADA2 pairwise distances were greater for NIST1 and NIST2 (which had greater variability in library size) compared to the JHU2 run which had the lowest pairwise distance.  
Mean pairwise distances were consistent across the JHU runs for Mothur and DADA2, suggesting they are better able to account for sequencing errors than other pipelines evaluated in this study. 
Conversely, the Deblur pipeline had the highest number of failed samples for JHU1, suggesting it is less robust to sequencing errors compared to the other pipelines (Table \@ref(tab:pipeCharTbl)).  

