---
title: "Discussion"
author: "Nate Olson"
date: "`r Sys.Date()`"
output:
  bookdown::pdf_document2:
    toc: FALSE
bibliography: ../bibliography.bib
---

Sequence error rate and variation in library size are just two sequencing characteristics that can negatively bias beta diversity analyses  [@McMurdie2014].
Ideally, bioinformatic pipelines can help differentiate true biological sequences from artifacts generated by sequencing errors [@callahan2016dada2] and normalization methods, such as rarefaction and cumulative sum scaling, can adjust for differences in library size [@paulson2013differentially].
However, the efficacy of these different pipelines and normalization techniques for microbiome datasets and how they affect study conclusions are not well characterized.
We compared the performance of six bioinformatic pipelines and nine normalization methods on mixture samples for four beta-diversity metrics, finding that these pipelines and methods vary significantly in their ability to identify and correct these biases.

We utilized a novel two-sample titration dataset of DNA extracts from five participants in a vaccine trial.
Individual titration series were generated for each participant, where DNA collected before exposure to pathogenic _E. coli_ were titrated into DNA samples collected after exposure.
These samples were processed with multiple levels of technical replication, including 16S rRNA PCR assays, sequencing libraries, and sequencing runs that were performed in duplicate at two independent laboratories. Our framework assessed three components:
(1) beta-diversity repeatability of PCR replicates,
(2) signal-to-noise analysis of the between to within-sample beta diversity of titration sets, and
(3) contribution of biological (subjects and exposure status) and technical factors (PCR replicates, sequencing labs, and runs) to beta-diversity.


When comparing PCR replicates for all sequencing runs, the QIIME de novo pipeline had high UniFrac values, but low weighted UniFrac values.
This is most likely due to the high proportion of singletons generated \@ref(tab:pipeCharTbl)).
A large number of singletons indicates that a pipeline is unable to group sequencing artifacts with true biological sequences.
The removal of singletons, a step included in many workflows such as the QIIME open-reference pipeline, can address this bias.
The differences we observed in weighted and unweighted Unifrac values also emphasize the importance of assessing multiple beta diversity metrics,
as each metric provides unique insight into community composition shifts.
Normalization methods generally improved beta-diversity repeatability,
with the exception of rarefying data to 15th quantile,
which resulted in higher beta-diversity between PCR replicates,
especially for QIIME pipelines, possibly due to large sample loss.
Count data normalized using TMM and RLE consistently had lower beta-diversity values between PCR replicates compared to un-normalized count data.

<!-- Use either exposure status or timepoint for unmixed samples -->

The biological signal magnitude was equal to the technical noise for un-normalized samples, highlighting the overall importance of normalization.
Rarefaction methods at lower subsampling depths generally increased the signal to noise ratio for unweighted metrics, especially for the DADA2 and Mothur pipelines.
Unexpectedly, most numeric normalization methods did not increase the signal-to-noise ratio for weighted metrics, and TMM and RLE normalization methods,
which showed the greatest similarity between PCR replicates,
decreased our ability to tease out the true biological indicators.


We finally evaluated the impact of difference sources of variability on pipeline and normalization methods by comparing diversity between biological samples and technical replicates.
For most pipelines and beta diversity metrics, normalizing the count data increased the difference in beta diversity between biological and technical replicates (Fig. \@ref(fig:techNormImpactWeighted)),
indicating a greater ability to detect community levels differences between treatment conditions.
Some metrics, namely rarefying to the 15th quantile, RLE, and TMM, frequently reduced the differences in beta-diversity between the biological to technical factors. Variation partitioning results were consistent with this conclusion (Fig. \@ref(fig:bioVtechVarpart)).

This study highlights the importance of rigorous methods evaluation of computational tools and datasets.
While we utilized six commonly cited bioinformatics pipelines, there are many different approaches and researchers should critically think about which is most appropriate for their own dataset.
We used default program parameters in our analyses to make our findings generally applicable.
However, we strongly advise researchers to have a good understanding of each step in their chosen pipeline, including what parameters are required and whether they should be changed to best fit data of interest.


Furthermore, this study shows the importance of normalizing microbiome count tables prior to beta-diversity analyses.
As the microbiome field is relatively young, many existing normalization approaches are adopted from methods created for other applications.
For instance, RLE and TMM normalization methods were initially developed for normalizing microarray and RNAseq data, not marker-gene sequence data.
 While these methods improve differential abundance analysis [@McMurdie2014], they may not appropriate for beta-diversity analysis.


## Conclusions
The results presented in this study can be used to help determine the appropriate bioinformatic pipeline and normalization method for a marker-gene survey beta diversity analysis.
The six pipelines evaluated in this study varied in their ability to distinguish sequencing artifacts from true biological sequences and these differences impacted the PCR replicate beta diversity repeatability.
Based on our study results we found Mothur and DADA2 to be more robust to lower quality sequence datasets.
Optimizing QIIME preprocessing methods may increase pipeline robustness to lower quality data.


Normalization can improve PCR replicate repeatability, but sometimes at the cost of decreasing the differences in beta diversity for biological relative to technical factors.
Our results indicate normalization methods developed for gene expression methods may not be appropriate for marker-gene survey beta diversity analysis.
For weighted metrics, we recommend normalizing counts using TSS and CSS improved assessment results or had no effect compared to unnormalized counts.
Rarefying count data improved unweighted metric results but higher rarefaction levels tended to perform worse than unnormalized data.
Rarefying counts lowers statistical power and therefore not it is not advisable when other normalization methods are available [@McMurdie2014].
As numeric normalization methods are not applicable to unweighted metrics, rarefying counts is the recommended normalization method.
To reduce the risk of the random subsampling step biasing beta diversity results bootstrap replicates can be used to validate results.


Bioinformatic pipelines combine multiple algorithms to convert raw sequence data into a count table which is subsequently used to test biological hypotheses.
Algorithm choice and parameters can significantly impact pipeline results.
The pipelines compared in this study were optimized using mock communities and benchmarked against other methods based on similarity in beta-diversity results [@bokulich2016mockrobiota].
The novel assessment framework and dataset presented here provides complementary methods for use in optimizing existing and benchmarking new pipelines and normalization methods.


<!--
Notes for discussion revision
- weighted phylogenetic lowest
  - Overall differences in beta diversity metrics are due to differences in how the four metrics measure community similarity.
  - For phylogenetic metric, the beta diversity tended to be lower compared to the taxonomic metrics.
  - This was due to low overall phylogenetic diversity, or the majority of the features being phylogenetically closely related.

- Data characteristics impact diversity metrics differently.
  - Higher unweighted PCR replicate beta diversity due to spurious features, sequencing artifacts.
  - Most obvious for QIIME _de novo_ where 2/3rds of the features are singletons.
  - Prevalence filtering as a way to address this issue.

- Unweighted metric values close to 1
  - In appropriate for these datasets
  - Value in unweighted metrics
  - Weighted metric values driven by high abundance organisms
  - Unweighted metrics provide insights into similarity between low abundance community members.
  - Phylogenetic metric, UniFrac, lower for most metrics - more robust to artifacts as they are likely similar to true sequences thus shorter branch lengths.

- consistent PCR replicate beta diversity across pipelines
  - taxonomic weighted
  - Implications for pipeline comparisons using beta diversity metrics?

- sequencing run differences
  - NIST runs consistently higher than JHU2 and JHU1 for Mothur and DADA2.
  - Potential reasoning for NIST being higher
  - Library specific effect
    - barcodes
    - run later
    - contaminants
  - NIST runs also larger variance for weighted taxa

- PCR replicates normalization
  - rarefaction level correlated with PCR replicate distance unweighted mtrics


- For the second component of our assessment we compare the signal-to-noise ratio between sequence data characteristics, bioinformatic pipeline, and normalization method.
  - Signal-to-noise around 1 for most pipelines and metrics.
  - signal to noise values limited by noise
    - when PCR rep beta 0.5 max signal-to-noise value is 2
  - Values around 1 indicate an inability to differentiate between samples.
  - Lack of different due to either high similarity between unmixed samples or data characteristics biasing diversity metrics
  - The assessment results varied by pipeline and diversity metric, with DADA2 and Mothur consistently out perform the other bioinformatic pipelines for weighted metrics.
  - Weighted greater than 1 for mothur and DADA2
  - JHU higher than NIST - consistent with PCR replicate results
  - Subject differences reflect differences in bacterial DNA proportions between unmixed samples [__REFERENCE CHAPTER__]
  - Normalization methods
    - TSS only method with no change or improved signal to noise
    - Rare2000 improved unweighted


PCR replicate beta diversity repeatability for some pipelines was more robust to low sequence quality and total abundance variation than others.
    - Mothur and DADA2 JHU1 consistent with JHU2
    - JHU1 higher compared to other runs for most QIIME pipelines and metrics
      - Deblur JHU1 failed
    - sequencing run differences
      - NIST runs consistently higher than JHU2 and JHU1 for Mothur and DADA2.
      - Potential reasoning for NIST being higher
      - Library specific effect
        - barcodes
        - run later
        - contaminants

  __PIPELINE RECOMENDATIONS__
  - Mothur and DADA2 more robust to low quality data, JHU1 results
  - QIIME pipelines
    - modifying pre-processing parameters
    - prevalence filtering

  __Normalization RECOMENDATIONS__
  - RLE and TMM not appropriate for beta diversity
  - Rarefaction statistically inadmisable
  - TSS and CSS for weighted metrics
  - Rarefaction only methods for unweighted metrics
  - Bootstrap rarefaction Bee reference

  __Metric Recomendations__
  - Higher signal to noise for weighted metrics
  - Care in interpreting results from unweighted metrics
    - prevalence filtering in addition to normalization may improve results
  - Alternative Hill number beta diversity
-->
