---
title: "Discussion"
author: "Nate Olson"
date: "`r Sys.Date()`"
output:
  bookdown::pdf_document2:
    toc: FALSE
bibliography: ../bibliography.bib
---

Diversity metric differences
- Why rarefaction is not statistically sound but should be used for unweighted metrics
  - Bootstrap replicates are a potential alternative
- difference in library size and total feature abundance
- pipeline recommendations
- normalization method recomendations

Sequence data characteristics, specifically sequence error rate and variation in library size can negatively impact beta diversity analysis (__REF__).
Bioinformatic pipelines are used to convert the raw sequence data to a count matrix, ideally differentiating true biological sequences from sequencing artifacts alleviating biases due to sequencing errors (__REF__).
Normalization methods, such as rarefying count data and cumulative sum scaling, are used to account for library size differences (__REF__).
For our assessment we compared the performance of six bioinformatic pipelines and nine normalization methods for four beta-diversity metrics.
The results from our study employing a novel assessment framework indicates that bioinformatic pipelines vary in their ability to alleviate biases in beta diversity due to sequencing errors and some normalization methods accentuate the biases in beta-diversity analysis due to library size differences.

Our assessment framework consisted of three components, (1) PCR beta-diversity repeatability, (2) signal-to-noise analysis, and (3) biological and technical factors contribution to beta-diversity variability.
Our assessment framework utilized a novel two-sample titration dataset with multiple levels of technical replication including  16S rRNA PCR, sequencing library, and sequencing run.
PCR replicates from within a seqeuncing were used to assess beta-diversity repeatability.
DNA extracts from five participants in a vaccine trial were used to for the two-sample titations.
Individual titration series were generated for each individual, where DNA collected before exposure to pathogenic _E. coli_ were titrated into DNA from samples collected after exposure.
And the titrations were used to assess signal-to-noise ratios.
Finally, the different sample types (subjects and exposure status) in conjunction with multiple sequencing runs were used to compare sources of variability.


Sequence quality and total abundance variation impact on PCR replicate beta diversity repeatability was pipeline and diversity metric dependent.
For the QIIME _de novo_ pipeline the PCR replicate unweighted UniFrac was high for all runs but low for weighted UniFrac.
We attributed the difference between QIIME _de novo_  the weighted and unweighted UniFrac metric to the high number of singletons in the dataset.
These sigletons indicate the pipelines inability to group sequencing artifacts with true biological sequences.
Singleton removal, a step included in the QIIME open-reference pipeline can be used to address this bias.
<!-- NIST runs
weighted and unweighted metrics
phylogenetic and taxonomic metrics

Overall differences in beta diversity metrics are due to differences in how the four metrics measure community similarity.
For phylogenetic metric, the beta diversity tended to be lower compared to the taxonomic metrics.
This was due to low overall phylogenetic diversity, or the majority of the features being phylogenetically closely related. -->

Normalization methods improved beta-diveristy repeatability, excluding rarefying data to 15th quantile, which resulted in higher beta-diversity between PCR replicates, especially for QIIME pipelines.
Count data normlized using TMM and RLE consistently had lower beta-diversity between PCR replicates compared to un-normalized count data.


For the second component of our assessment we compare the signal-to-noise ratio between sequence data characteristics, bioinformatic pipeline, and normalization method.
The assessment results varied by pipeline and diversity metric, with DADA2 and mothur consistently out performing the other bioinformatic pipelines for weighted metrics.
__MORE__


To evaluate the impact of difference sources of variability on bioinformatic pipeline and normalization methods we compared beta diversity between different biological samples and technical replicates, including PCR replicates and sequencing runs.
Differences in beta diversity between biological samples and technical replicates varied by diversity metric, pipeline, and normalization method.
For most pipelines and beta diversity metrics, normalizing the count data increased the difference in beta diversity between biological and technical replicates, resulting in a greater ability to detect beta diversity differences between communities from different treatments.
However, some metrics, namely rarefying to the 15th quantile, RLE, and TMM, frequently reduced the differences in beta-diversity between the biological to technical factors.
Variation partitioning results were consistent with this conclusion.
RLE and TMM were developed for normalizing microarray and RNAseq data and not marker-gene sequence data.
While these normalization methods have been show to be useful for differential abundance analysis, they are not appropriate for beta-diversity analysis.

## Conclusions
The results presented in this study can be used to help determine the appropriate bioinformatic pipeline and normalization method for a marker-gene survey beta diversity analysis.
The six pipelines evaluated in this study varied in their ability to distinguish sequencing artifacts from true biological sequences and these differences impacted the PCR replicate beta diversity repeatability.
Normalization can help improve repeatability, but sometimes at the cost of decreasing the differences in beta diversity for biological relative to technical factors.
Based on our study resutls we found that Mothur and DADA2 are more robust to lower quality datasets.
Furthermore, normalization methods can improve our ability to detect true biological differences though normalization methods developed for gene expression methods may not be appropriate for marker-gene survey beta diversity analysis.

Bioinformatic pipelines combine multiple algorithms to convert raw sequence data into a count table which are subsequently used to test biological hypotheses.
Algorithm choice and parameters can significantly impact pipeline results.
The pipelines compared in this study were optimized using mock communities and benchmarked against other methods based on similarity in beta-diversity results [@bokulich2016mockrobiota].
The novel assessment framework and dataset presented here provides complementary methods for use in optimizeing existing and benchmarking new pipelines and normalization methods.
