---
title: "Discussion"
author: "Nate Olson"
date: "`r Sys.Date()`"
output:
  bookdown::pdf_document2:
    toc: FALSE
bibliography: ../bibliography.bib
---

Sequence data characteristics, specifically sequence error rate and variation in library size can negatively impact beta diversity analysis (__REF__).
Bioinformatic pipelines are used to convert the raw sequence data to a count matrix, ideally differentiating true biological sequences from sequencing artifacts alleviating biases due to sequencing errors.
Normalization methods, such as rarefying count data and cumulative sum scaling, are used to account for library size differences.
For our assessment we compared the performance of six bioinformatic pipelines and nine normalization methods for four beta-diversity metrics.
The results from our study employing a novel dataset and assessment framework indicate that bioinformatic pipelines vary in their ability to alleviate biases in beta diversity due to sequencing errors and some normalization methods accentuate the biases in beta diversity analysis due to library size differences.

Our assessment framework consisted of three components, (1) PCR beta-diversity metric repeatability, (2) signal-to-noise analysis, (3) difference in variation between biological factors and technical replicates.
Our assessment framework utilized a novel two-sample titration dataset with multiple levels of technical replication including  16S rRNA PCR, sequencing library, and sequencing run.
Multiple PCR replicates were used to assess beta-diversity repeatability.
The titrations were used to assess signal-to-noise ratios.
Finally, the different sample types (trial participants and exposure status) in conjunction with multiple sequencing runs were used to compare sources of variability.


Using PCR replicate beta diversity for the four sequencing runs we showed how sequence quality and total abundance variation impacts PCR replicate beta diversity repeatability was pipeline and diversity metric dependent.
For the QIIME _de novo_ pipeline the PCR replicate unweighted UniFrac was high for all runs but low for weighted UniFrac.
We attributed the difference between QIIME _de novo_  the weighted and unweighted UniFrac metric to the high number of singletons in the dataset.
These sigletons indicate the in ability of the pipeline to group sequencing artifacts with true biological sequences.
Singleton removal, a step included in the QIIME open-reference pipeline addresses this bias.
Sequence data from JHU run 1, which had lower error rate and total abundance relative to the other sequencing runs had consistently better repeatability across pipelines and diversity metrics.
Normalization methods improved repeatability, excluding rarefying data to 15th quantile, which decreased repeatability especially for QIIME pipelines. TMM improved weighted beta diversity repeatability for NIST datasets, greater variability in library size.
While it is important to reduce the beta-diversity between technical replicates, it is more important to be able to detect true differences between biological samples.
To detect differences between biological samples, sample dissimilarity due to biological factors must be greater than sample dissimilarity due to technical variability or noise.


For the second component of our assessment we compare the signal-to-noise ratio between sequence data characteristics, bioinformatic pipeline, and normalization method.
The assessment results varied by pipeline and diversity metric, with DADA2 and mothur consistently out performing the other bioinformatic pipelines for weighted metrics.
__MORE__


To evaluate the impact of difference sources of variability on bioinformatic pipeline and normalization methods we compared the beta diversity between different biological samples and technical replicates, including PCR replicates and sequencing runs.
Differences in beta diversity between biological samples and technical replicates varied by diversity metric, pipeline, and normalization method.
Overall differences in beta diversity metrics are due to differences in how the four metrics measure community similarity.
For phylogenetic metric, the beta diversity tended to be lower compared to the taxonomic metrics.
This was due to low overall phylogenetic diversity, or the majority of the features being phylogenetically closely related.
For most pipelines and beta diversity metrics, normalizing the count data increased the difference in beta diversity between biological and technical replicates, resulting in a greater ability to detect beta diversity differences between communities from different treatments.
However, some metrics, namely rarefying to the 15th quantile, RLE, and TMM, frequently reduced the difference between the biological signal and noise due to technical variability.
Variation partitioning results were consistent with this conclusion.
RLE and TMM were developed for normalizing microarray and and RNAseq data and not marker-gene sequence data.
While these normalization methods have been show to be useful for differential abundance analysis, they are not appropriate for beta-diversity analysis.

## Conclusions
The results presented in this study can be used to help determine the appropriate bioinformatic pipeline and normalization method for a marker-gene survey beta diversity analysis.
The six pipelines evaluated in this study varied in their ability to distinguish sequencing artifacts from true biological sequences.
These differences impacted the beta diversity repeatability.
Normalization can help improve repeatability, but sometimes at the cost of decreasing the difference between biological signal and technical variability.
Mothur and DADA2 are more robust to lower quality datasets.
Normalization methods can improve ability to detect true biological signal though normalization methods developed for gene expression methods may not be appropriate.

Bioinformatic pipelines combine multiple algorithms to convert the raw sequence data into a count table for use in statistical analysis.
The algorithm choice and parameters can significantly impact pipeline results.
The pipelines compared in this study were optimized using mock communities and benchmarked against other methods based on similarity in beta-diversity results [@bokulich2016mockrobiota].
The novel assessment framework and dataset presented here provides complementary methods for use in optimizeing existing and benchmarking new pipelines and normalization methods.
