---
title: "Methods"
author: "Nate Olson"
date: "`r Sys.Date()`"
output:
  bookdown::pdf_document2:
    toc: FALSE
bibliography: ../bibliography.bib
---

Our assessment framework utilizes a dataset of DNA mixtures from five vaccine trial participants described in Section \@ref(two-sample-titration-design).
DNA was extracted from stool collected from five individuals (subjects) before and after exposure to pathogenic _Escherichia coli_ (timepoints).
The pre- and post-exposure DNA was mixed following a $log_2$ two-sample titration mixture design, resulting in a set of samples with varying levels of similarity.
The microbial community in the unmixed pre- and post-exposure samples and titrations were measured using 16S rRNA marker-gene sequencing.
Four technical replicates of each were generated during the 16S rRNA PCR amplification process.
Technical replicates of each PCR were sent to two independent laboratories (JHU and NIST) for sequencing.


Sequencing libraries were prepared at the independent laboratories using the same protocol (ILLUMINA).
Resulting libraries were sequenced twice at each laboratory, resulting in four sequence datasets with varying sequence quality and library sizes.
The first JHU run PhiX error rate was higher than expected and the instrument was re-calibrated by the manufacturer, resulting in improved quality scores for the second run.
The first run at NIST generated lower total throughput than expected, so the pool library for the second run was re-optimized and generated a dataset with increased throughput and lower sample to sample read count variability.
No template controls were also sequenced for quality control and did not reveal any significant reagent contamination.
Sequence data characterization was performed using the savR [@savR] and ShortRead Bioconductor R packages [@ShortRead].


### Bioinformatic Pipelines
Data from the four sequencing runs were processed using six bioinformatic pipelines, including the QIIME open reference, closed reference, _de novo_, and Deblur pipelines, as well as the Mothur _de novo_ pipeline and DADA2 sequence inference pipeline.
The code used to run the bioinformatic pipelines is available at https://github.com/nate-d-olson/mgtst_pipelines/, on the multirun branch.
Pre-processing and feature detection methods vary by pipeline.
The Mothur pipeline uses the OptiClust algorithm for _de novo_ clustering [@westcott2017opticlust].
Pre-processing includes merging and quality filtering paired-end reads followed by aligning sequences to the SILVA reference alignment [@schloss2009introducing].
Taxonomic classification was performed using the RDP Bayesian classifier [@wang2007naive] implemented in Mothur.
The phylogenetic tree was constructed in Mothur using the clearcut algorithm [@Sheneman2006].
Mothur version 1.39.3 (https://www.mothur.org) and SILVA release version 119 reference alignment and RDP the mothur formatted version of the RDP 16S rRNA database release version 10 [@Cole2014].


The DADA2 big data protocol for DADA2 versions 1.4 or later was followed (https://benjjneb.github.io/dada2/bigdata.html), except for read length trimming parameters and primer trimming.
Forward and reverse primers were trimmed using cutadapt version 1.14 (https://cutadapt.readthedocs.io/en/stable/) [@Martin2011].
The forward and reverse reads were trimmed to 260 and 200 bp respectively.
Read trimming positions were defined based on read quality score distributions, maximizing the overlap region between the forward and reverse read while minimizing the inclusion of low-quality sequence data.
The pipeline was run using DADA2 version 1.6.0 [@callahan2016dada2] and formatted SILVA database version 128 trainset provided by the DADA2 developers [@benjamin_callahan_2017_824551].
Taxonomic classification was performed using the DADA2 implementation of the RDP Bayesian classifier [@wang2007naive].
The phylogenetic tree was generated following methods in [@callahan2016] using the DECIPHER R package for multiple sequence alignment [@DECIPHER] and the phangorn R package for tree construction [@phangorn].


The QIIME pipelines all used the same merged paired-end, quality filtered set of sequences [@Caporaso2010].
UCLUST alogrithm (version v1.2.22q) was used for clustering and taxonomic assignment against the Greengenes database version 13.8 97\% similarity OTUs [@edgar2010search;@McDonald2012-ro].
Phylogenetic trees were constructed using FastTree, and a multiple sequence alignment generated using pyNAST and the Greengenes reference alignment [@Caporaso2010;@Price2010].
Both open and closed reference pipelines used the Greengenes 97\% similarity database for reference clustering.
Additionally, sequence variants were inferred from the QIIME merged and quality-filtered sequences using Deblur (version 1.0.3) [@Amir2017].
Phylogenetic tree construction methods used for the other QIIME pipelines were also used for the Deblur pipeline.

### Normalization Methods and Beta-Diversity Metrics
Normalization methods are used to account for between sample differences in feature total abundance.
Rarefaction, subsampling counts without replacement to an even abundance, is a commonly used normalization method in macro-ecology and 16S rRNA marker-gene surveys [@Gotelli2001;@Hughes2005].
We rarefied samples to four levels; 2000, 5000, and 10000 total reads per sample, and to the total abundance of the 15th percentile.
Rarefaction levels were selected based on values used in published studies [@Thompson2017] and other comparison studies [@Weiss2017;@McMurdie2014].
Rarified count data were analyzed using both weighted and unweighted beta-diversity metrics.
Numeric normalization methods include those previously developed for normalizing microarray and RNAseq data, such as upper quartile (UQ), trimmed mean of M values (TMM), and relative log expression [@Robinson2010;@McCarthy2012], and those that are commonly used to normalize 16S rRNA marker-gene survey, such as cumulative sum scaling (CSS) [@paulson2013differential] and total sum scaling (proportions, TSS).
Numeric normalization methods were used for weighted metrics, as they do not impact unweighted metric results.

Weighted and unweighted phylogenetic and taxonomic beta-diversity metrics were compared. Beta-diversity metrics were calculated using phyloseq version 1.22.3 [@McMurdie2013]. Weighted and unweighted UniFrac phylogenetic beta-diversity metrics were calculated using the phyloseq implementation of FastUniFrac [@McMurdie2013;@Hamady2010]. For feature-level beta-diversity assessment, the Bray-Curtis weighted, and Jaccard unweighted metrics were used [@Bray1957;@Jaccard1912].

### Beta-Diversity Assessment
Standard linear models were used to test for significance using the R `lm` function.
Mixed effects models, used to take into account repeated measures, were fit using the R `lmer` function in the lme4 package [@lme4].
Model fit was evaluated based on model statistics, AIC, BIC, and logLik, as well as diagnostic plots.
Tukey Honest Significant Differences test was used for multiple comparison testing using the `TukeyHSD` function.
The source code for all analysis is available at https://github.com/nate-d-olson/diversity_assessment.

#### Beta-Diversity Repeatability

Beta-diversity repeatability was evaluated for the different pipelines across sequencing runs.
Here we define repeatability as the median beta diversity between PCR replicates.
The unnormalized count data was used to characterize the baseline beta-diversity repeatability for the different pipeline and sequencing runs.
Linear models were used to quantify differences between pipelines and across the four sequencing runs for the diversity metrics.
Data from a sequencing run NIST1 were used to evaluate normalization method impact on PCR replicate beta-diversity.
To quantify normalization method impact, independent linear models were fit for each pipeline and diversity metric.


<!-- - R model equation  `mean_dist~pipe*seq_run_id`, `mean_dist` is the mean beta-diversity between the four PCR replicates, `pipe` is bioinformatic pipeline, and `seq_run_id` is sequencing run number (note does not account for lab effect).
- Linear model used to quantify normalization method impact on PCR repeatability.
    - Independent models fit for each diversity metric-pipeline combination.
    - R model equation `mean_dist~normalization`.
-->

#### Beta Diversity Signal to Noise Ratio

Next, we evaluated the signal-to-noise ratio for the different pipelines across sequencing runs by comparing pre-exposure samples to other samples in the titration series.
Signal was measured as the median beta-diversity between samples were compared (Fig. \@ref(fig:experimentalDesign)).
Noise was measured as the median PCR replicate beta-diversity within the compared samples.
A weighted average of the signal-to-noise ratio was calculated as the area under the curve (using the `trapz` function) of the signal-to-noise ratio and the proportion of pre-exposure DNA in the sample being compared [@pracma].
Independent linear models were fit for each diversity metric to quantify differences in the signal-to-noise ratio between sequencing runs and pipelines.
A mixed-effects linear model was then used to quantify normalization method impact on the signal-to-noise ratio using data from NIST1 with subject as a random effect.
Independent mixed effects linear models were fit for each pipeline and diversity metric.

<!-- linear model was then used  and diversity metric `log10(auc) ~ pipe + seq_run_num + biosample_id`.
norm effect `log10(auc) ~ method + (1 | biosample_id)`
-->


#### Beta Diversity Between Individuals and Treatments

To quantify the contribution of biological and technical variability to total variability the distribution of beta diversity metrics were compared between subjects, within subject and between conditions (pre- and post-exposure), and different types of technical replicates.
A linear model was used to quantify differences in beta diversity between biological and technical sources of variability.
We then used variation partitioning [@borcard1992partialling] to quantify technical and biological factor's contribution to the total observed variation.
Variation partition was calculated using the Vegan R package [@vegan].
Distance-based redundancy analysis (dbRDA) was used to identify significant sources of variation [@vegan].

<!-- - R model equation `log(value) ~ variation + variation_label`, `value` is the beta diversity between replicates, `variation_label` is the variation source, and `variation` is the type of variation.
- Variation labels
    - btw_subject_w/in_time: between subjects within treatment (pre- and post-exposure), within sequencing run
    - w/in_subj_btw_time: within subject between treatments, within sequencing run
    - btw_time: across subjects between treatments, within sequencing run
    - w/in_lab_pcr: with subject, treatment and lab, between sequencing runs?? (check)
    - w/in_lab_runs: within subject, treatment, lab, and sequencing runs (4 pcr replicates)
    - btw_labs: within subject, treatment, across sequencing labs -->
